{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ae48563b-8543-45f3-86e9-ca5639e629f5",
   "metadata": {},
   "source": [
    "# Setup prompts and interact with IBM BAM\n",
    "\n",
    "In this notebook, we will interact with IBM models and setup input prompts to get started with the api docs generator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "57f27b93-4b22-4e10-848e-649c3468513c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:46.766251Z",
     "iopub.status.busy": "2023-11-21T15:30:46.765962Z",
     "iopub.status.idle": "2023-11-21T15:30:47.728130Z",
     "shell.execute_reply": "2023-11-21T15:30:47.727503Z",
     "shell.execute_reply.started": "2023-11-21T15:30:46.766192Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/app-root/lib64/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import json\n",
    "import re\n",
    "import pandas as pd\n",
    "from dotenv import load_dotenv\n",
    "from genai.credentials import Credentials\n",
    "from genai.model import Model\n",
    "from genai.schemas import GenerateParams"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "366a1860-7cbc-4428-8c3a-eaced387d1ae",
   "metadata": {},
   "source": [
    "### Setup BAM API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "544a72e2-c16d-48af-9fc6-2dbb0c784281",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.729982Z",
     "iopub.status.busy": "2023-11-21T15:30:47.729570Z",
     "iopub.status.idle": "2023-11-21T15:30:47.734380Z",
     "shell.execute_reply": "2023-11-21T15:30:47.733838Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.729958Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# make sure you have a .env file in the root folder with genaikey and genaiapi\n",
    "load_dotenv()\n",
    "api_key = os.getenv(\"GENAI_KEY\", None)\n",
    "api_endpoint = os.getenv(\"GENAI_API\", None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2ce3d4e2-0c2d-4034-bb70-494dffe2ad15",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.735402Z",
     "iopub.status.busy": "2023-11-21T15:30:47.735156Z",
     "iopub.status.idle": "2023-11-21T15:30:47.739603Z",
     "shell.execute_reply": "2023-11-21T15:30:47.739091Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.735383Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# creds object\n",
    "creds = Credentials(api_key=api_key, api_endpoint=api_endpoint)\n",
    "\n",
    "# Instantiate parameters for text generation\n",
    "params = GenerateParams(\n",
    "    decoding_method=\"sample\",\n",
    "    max_new_tokens=1024,\n",
    "    temperature=0.7\n",
    ")\n",
    "\n",
    "model = \"meta-llama/llama-2-7b\"\n",
    "MAX_TOKEN_LENGTH = 4096\n",
    "# Instantiate a model proxy object to send your requests\n",
    "chosen_model = Model(model, params=params, credentials=creds)\n",
    "\n",
    "# Instantiate a paramater dictionary object\n",
    "params_dict = vars(params)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee0ea013-d45f-463c-bbf1-a2c756c5c631",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "565ae642-87d9-423b-9127-9b79e7012dbe",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.740556Z",
     "iopub.status.busy": "2023-11-21T15:30:47.740309Z",
     "iopub.status.idle": "2023-11-21T15:30:47.743842Z",
     "shell.execute_reply": "2023-11-21T15:30:47.743331Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.740537Z"
    }
   },
   "outputs": [],
   "source": [
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "13e06fe0-f8da-440a-b609-48ced0161095",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.744818Z",
     "iopub.status.busy": "2023-11-21T15:30:47.744557Z",
     "iopub.status.idle": "2023-11-21T15:30:47.749858Z",
     "shell.execute_reply": "2023-11-21T15:30:47.749348Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.744799Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load markdown data\n",
    "dataset_path = \"../../data/raw/markdown_data.json\"\n",
    "with open(dataset_path, 'r') as f:\n",
    "\t\tmarkdown = json.load(f)\n",
    "\n",
    "# load nested data\n",
    "dataset_path = \"../../data/raw/nested_data.json\"\n",
    "with open(dataset_path, 'r') as f:\n",
    "\t\tdata = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "29bf1d44-1285-4fb0-af30-203a306caf23",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.750835Z",
     "iopub.status.busy": "2023-11-21T15:30:47.750586Z",
     "iopub.status.idle": "2023-11-21T15:30:47.758874Z",
     "shell.execute_reply": "2023-11-21T15:30:47.758350Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.750816Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['errors', 'oidc', 'sign', 'transparency', 'verify'])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markdown.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0d089c18-cae1-4538-8ae5-3ac61b377988",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.761123Z",
     "iopub.status.busy": "2023-11-21T15:30:47.760855Z",
     "iopub.status.idle": "2023-11-21T15:30:47.764737Z",
     "shell.execute_reply": "2023-11-21T15:30:47.764235Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.761104Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['errors', 'oidc', 'sign', 'transparency', 'verify_models', 'verify_policy', 'verify_verifier'])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "840fbc65-e600-454e-860c-6627e7a02cf9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.765704Z",
     "iopub.status.busy": "2023-11-21T15:30:47.765460Z",
     "iopub.status.idle": "2023-11-21T15:30:47.768586Z",
     "shell.execute_reply": "2023-11-21T15:30:47.768078Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.765686Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "code_snippet = data[\"oidc\"][\"code\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6b660f75-8d0d-40c0-bc6e-df1e9c9512ee",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.769533Z",
     "iopub.status.busy": "2023-11-21T15:30:47.769296Z",
     "iopub.status.idle": "2023-11-21T15:30:47.773112Z",
     "shell.execute_reply": "2023-11-21T15:30:47.772611Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.769515Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['oidc.py'])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "code_snippet.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e47d50b1-e95a-44f7-b7ed-9fb1b39fb0c1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.774058Z",
     "iopub.status.busy": "2023-11-21T15:30:47.773802Z",
     "iopub.status.idle": "2023-11-21T15:30:47.776824Z",
     "shell.execute_reply": "2023-11-21T15:30:47.776325Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.774040Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "code_snippet = code_snippet[\"oidc.py\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e027d7b2-e84b-4aec-bbd2-fc705febb7d9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.777780Z",
     "iopub.status.busy": "2023-11-21T15:30:47.777535Z",
     "iopub.status.idle": "2023-11-21T15:30:47.781324Z",
     "shell.execute_reply": "2023-11-21T15:30:47.780798Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.777761Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Copyright 2022 The Sigstore Authors\n",
      "#\n",
      "# Licensed under the Apache License, Version 2.0 (the \"License\");\n",
      "# you may not use this file except in compliance with the License.\n",
      "# You may obtain a copy of the License at\n",
      "#\n",
      "#      http://www.apache.org/licenses/LICENSE-2.0\n",
      "#\n",
      "# Unless required by applicable law or agreed to in writing, software\n",
      "# distributed under the License is distributed on an \"AS IS\" BASIS,\n",
      "# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
      "# See the License for the specific language governing permissions and\n",
      "# limitations under the License.\n",
      "\n",
      "\"\"\"\n",
      "API for retrieving OIDC tokens.\n",
      "\"\"\"\n",
      "\n",
      "from __future__ import annotations\n",
      "\n",
      "import logging\n",
      "import sys\n",
      "import time\n",
      "import urllib.parse\n",
      "import webbrowser\n",
      "from datetime import datetime, timezone\n",
      "from typing import NoReturn, Optional, cast\n",
      "\n",
      "import id\n",
      "import jwt\n",
      "import requests\n",
      "from pydantic import BaseModel, StrictStr\n",
      "\n",
      "from sigstore.errors import Error, NetworkError\n",
      "\n",
      "DEFAULT_OAUTH_ISSUER_URL = \"https://oauth2.sigstore.dev/auth\"\n",
      "STAGING_OAUTH_ISSUER_URL = \"https://oauth2.sigstage.dev/auth\"\n",
      "\n",
      "# See: https://github.com/sigstore/fulcio/blob/b2186c0/pkg/config/config.go#L182-L201\n",
      "_KNOWN_OIDC_ISSUERS = {\n",
      "    \"https://accounts.google.com\": \"email\",\n",
      "    \"https://oauth2.sigstore.dev/auth\": \"email\",\n",
      "    \"https://oauth2.sigstage.dev/auth\": \"email\",\n",
      "    \"https://token.actions.githubusercontent.com\": \"sub\",\n",
      "}\n",
      "_DEFAULT_AUDIENCE = \"sigstore\"\n",
      "\n",
      "\n",
      "class _OpenIDConfiguration(BaseModel):\n",
      "    \"\"\"\n",
      "    Represents a (subset) of the fields provided by an OpenID Connect provider's\n",
      "    `.well-known/openid-configuration` response, as defined by OpenID Connect Discovery.\n",
      "\n",
      "    See: <https://openid.net/specs/openid-connect-discovery-1_0.html>\n",
      "    \"\"\"\n",
      "\n",
      "    authorization_endpoint: StrictStr\n",
      "    token_endpoint: StrictStr\n",
      "\n",
      "\n",
      "# See: https://github.com/sigstore/fulcio/blob/b2186c0/pkg/config/config.go#L182-L201\n",
      "_KNOWN_OIDC_ISSUERS = {\n",
      "    \"https://accounts.google.com\": \"email\",\n",
      "    \"https://oauth2.sigstore.dev/auth\": \"email\",\n",
      "    \"https://oauth2.sigstage.dev/auth\": \"email\",\n",
      "    \"https://token.actions.githubusercontent.com\": \"sub\",\n",
      "}\n",
      "DEFAULT_AUDIENCE = \"sigstore\"\n",
      "\n",
      "\n",
      "class ExpiredIdentity(Exception):\n",
      "    \"\"\"An error raised when an identity token is expired.\"\"\"\n",
      "\n",
      "\n",
      "class IdentityToken:\n",
      "    \"\"\"\n",
      "    An OIDC \"identity\", corresponding to an underlying OIDC token with\n",
      "    a sensible subject, issuer, and audience for Sigstore purposes.\n",
      "    \"\"\"\n",
      "\n",
      "    def __init__(self, raw_token: str) -> None:\n",
      "        \"\"\"\n",
      "        Create a new `IdentityToken` from the given OIDC token.\n",
      "        \"\"\"\n",
      "\n",
      "        self._raw_token = raw_token\n",
      "\n",
      "        # NOTE: The lack of verification here is intentional, and is part of\n",
      "        # Sigstore's verification model: clients like sigstore-python are\n",
      "        # responsible only for forwarding the OIDC identity to Fulcio for\n",
      "        # certificate binding and issuance.\n",
      "        try:\n",
      "            self._unverified_claims = jwt.decode(\n",
      "                raw_token,\n",
      "                options={\n",
      "                    \"verify_signature\": False,\n",
      "                    \"verify_aud\": True,\n",
      "                    \"verify_iat\": True,\n",
      "                    \"verify_exp\": True,\n",
      "                    # These claims are required by OpenID Connect, so\n",
      "                    # we can strongly enforce their presence.\n",
      "                    # See: https://openid.net/specs/openid-connect-basic-1_0.html#IDToken\n",
      "                    \"require\": [\"aud\", \"sub\", \"iat\", \"exp\", \"iss\"],\n",
      "                },\n",
      "                audience=DEFAULT_AUDIENCE,\n",
      "                # NOTE: This leeway shouldn't be strictly necessary, but is\n",
      "                # included to preempt any (small) skew between the host\n",
      "                # and the originating IdP.\n",
      "                leeway=5,\n",
      "            )\n",
      "        except Exception as exc:\n",
      "            raise IdentityError(\n",
      "                \"Identity token is malformed or missing claims\"\n",
      "            ) from exc\n",
      "\n",
      "        self._iss: str = self._unverified_claims[\"iss\"]\n",
      "        self._nbf: int | None = self._unverified_claims.get(\"nbf\")\n",
      "        self._exp: int = self._unverified_claims[\"exp\"]\n",
      "\n",
      "        # Fail early if this token isn't within its validity period.\n",
      "        if not self.in_validity_period():\n",
      "            raise IdentityError(\"Identity token is not within its validity period\")\n",
      "\n",
      "        # When verifying the private key possession proof, Fulcio uses\n",
      "        # different claims depending on the token's issuer.\n",
      "        # We currently special-case a handful of these, and fall back\n",
      "        # on signing the \"sub\" claim otherwise.\n",
      "        identity_claim = _KNOWN_OIDC_ISSUERS.get(self.issuer)\n",
      "        if identity_claim is not None:\n",
      "            if identity_claim not in self._unverified_claims:\n",
      "                raise IdentityError(\n",
      "                    f\"Identity token is missing the required {identity_claim!r} claim\"\n",
      "                )\n",
      "\n",
      "            self._identity = str(self._unverified_claims.get(identity_claim))\n",
      "        else:\n",
      "            try:\n",
      "                self._identity = str(self._unverified_claims[\"sub\"])\n",
      "            except KeyError:\n",
      "                raise IdentityError(\n",
      "                    \"Identity token is missing the required 'sub' claim\"\n",
      "                )\n",
      "\n",
      "        # This identity token might have been retrieved directly from\n",
      "        # an identity provider, or it might be a \"federated\" identity token\n",
      "        # retrieved from a federated IdP (e.g., Sigstore's own Dex instance).\n",
      "        # In the latter case, the claims will also include a `federated_claims`\n",
      "        # set, which in turn should include a `connector_id` that reflects\n",
      "        # the \"real\" token issuer. We retrieve this, despite technically\n",
      "        # being an implementation detail, because it has value to client\n",
      "        # users: a client might want to make sure that its user is identifying\n",
      "        # with a *particular* IdP, which means that they need to pierce the\n",
      "        # federation layer to check which IdP is actually being used.\n",
      "        self._federated_issuer: str | None = None\n",
      "        federated_claims = self._unverified_claims.get(\"federated_claims\")\n",
      "        if federated_claims is not None:\n",
      "            if not isinstance(federated_claims, dict):\n",
      "                raise IdentityError(\n",
      "                    \"unexpected claim type: federated_claims is not a dict\"\n",
      "                )\n",
      "\n",
      "            federated_issuer = federated_claims.get(\"connector_id\")\n",
      "            if federated_issuer is not None:\n",
      "                if not isinstance(federated_issuer, str):\n",
      "                    raise IdentityError(\n",
      "                        \"unexpected claim type: federated_claims.connector_id is not a string\"\n",
      "                    )\n",
      "\n",
      "                self._federated_issuer = federated_issuer\n",
      "\n",
      "    def in_validity_period(self) -> bool:\n",
      "        \"\"\"\n",
      "        Returns whether or not this `Identity` is currently within its self-stated validity period.\n",
      "\n",
      "        NOTE: As noted in `Identity.__init__`, this is not a verifying wrapper;\n",
      "        the check here only asserts whether the *unverified* identity's claims\n",
      "        are within their validity period.\n",
      "        \"\"\"\n",
      "\n",
      "        now = datetime.now(timezone.utc).timestamp()\n",
      "\n",
      "        if self._nbf is not None:\n",
      "            return self._nbf <= now < self._exp\n",
      "        else:\n",
      "            return now < self._exp\n",
      "\n",
      "    @property\n",
      "    def identity(self) -> str:\n",
      "        \"\"\"\n",
      "        Returns this `IdentityToken`'s underlying \"subject\".\n",
      "\n",
      "        Note that this is **not** always the `sub` claim in the corresponding\n",
      "        identity token: depending onm the token's issuer, it may be a *different*\n",
      "        claim, such as `email`. This corresponds to the Sigstore ecosystem's\n",
      "        behavior, e.g. in each issued certificate's SAN.\n",
      "        \"\"\"\n",
      "        return self._identity\n",
      "\n",
      "    @property\n",
      "    def issuer(self) -> str:\n",
      "        \"\"\"\n",
      "        Returns a URL identifying this `IdentityToken`'s issuer.\n",
      "        \"\"\"\n",
      "        return self._iss\n",
      "\n",
      "    @property\n",
      "    def expected_certificate_subject(self) -> str:\n",
      "        \"\"\"\n",
      "        Returns a URL identifying the **expected** subject for any Sigstore\n",
      "        certificate issued against this identity token.\n",
      "\n",
      "        The behavior of this field is slightly subtle: for non-federated\n",
      "        identity providers (like a token issued directly by Google's IdP) it\n",
      "        should be exactly equivalent to `IdentityToken.issuer`. For federated\n",
      "        issuers (like Sigstore's own federated IdP) it should be equivalent to\n",
      "        the underlying federated issuer's URL, which is kept in an\n",
      "        implementation-defined claim.\n",
      "\n",
      "        This attribute exists so that clients who wish to inspect the expected\n",
      "        subject of their certificates can do so without relying on\n",
      "        implementation-specific behavior.\n",
      "        \"\"\"\n",
      "        if self._federated_issuer is not None:\n",
      "            return self._federated_issuer\n",
      "\n",
      "        return self.issuer\n",
      "\n",
      "    def __str__(self) -> str:\n",
      "        \"\"\"\n",
      "        Returns the underlying OIDC token for this identity.\n",
      "\n",
      "        That this token is secret in nature and **MUST NOT** be disclosed.\n",
      "        \"\"\"\n",
      "        return self._raw_token\n",
      "\n",
      "\n",
      "class IssuerError(Exception):\n",
      "    \"\"\"\n",
      "    Raised on any communication or format error with an OIDC issuer.\n",
      "    \"\"\"\n",
      "\n",
      "    pass\n",
      "\n",
      "\n",
      "class Issuer:\n",
      "    \"\"\"\n",
      "    Represents an OIDC issuer (IdP).\n",
      "    \"\"\"\n",
      "\n",
      "    def __init__(self, base_url: str) -> None:\n",
      "        \"\"\"\n",
      "        Create a new `Issuer` from the given base URL.\n",
      "\n",
      "        This URL is used to locate an OpenID Connect configuration file,\n",
      "        which is then used to bootstrap the issuer's state (such\n",
      "        as authorization and token endpoints).\n",
      "        \"\"\"\n",
      "        oidc_config_url = urllib.parse.urljoin(\n",
      "            f\"{base_url}/\", \".well-known/openid-configuration\"\n",
      "        )\n",
      "\n",
      "        try:\n",
      "            resp: requests.Response = requests.get(oidc_config_url, timeout=30)\n",
      "        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "            raise NetworkError from exc\n",
      "\n",
      "        try:\n",
      "            resp.raise_for_status()\n",
      "        except requests.HTTPError as http_error:\n",
      "            raise IssuerError from http_error\n",
      "\n",
      "        try:\n",
      "            # We don't generally expect this to fail (since the provider should\n",
      "            # return a non-success HTTP code which we catch above), but we\n",
      "            # check just in case we have a misbehaving OIDC issuer.\n",
      "            self.oidc_config = _OpenIDConfiguration.model_validate(resp.json())\n",
      "        except ValueError as exc:\n",
      "            raise IssuerError(f\"OIDC issuer returned invalid configuration: {exc}\")\n",
      "\n",
      "    @classmethod\n",
      "    def production(cls) -> Issuer:\n",
      "        \"\"\"\n",
      "        Returns an `Issuer` configured against Sigstore's production-level services.\n",
      "        \"\"\"\n",
      "        return cls(DEFAULT_OAUTH_ISSUER_URL)\n",
      "\n",
      "    @classmethod\n",
      "    def staging(cls) -> Issuer:\n",
      "        \"\"\"\n",
      "        Returns an `Issuer` configured against Sigstore's staging-level services.\n",
      "        \"\"\"\n",
      "        return cls(STAGING_OAUTH_ISSUER_URL)\n",
      "\n",
      "    def identity_token(  # nosec: B107\n",
      "        self,\n",
      "        client_id: str = \"sigstore\",\n",
      "        client_secret: str = \"\",\n",
      "        force_oob: bool = False,\n",
      "    ) -> IdentityToken:\n",
      "        \"\"\"\n",
      "        Retrieves and returns an `IdentityToken` from the current `Issuer`, via OAuth.\n",
      "\n",
      "        This function blocks on user interaction.\n",
      "\n",
      "        The `force_oob` flag controls the kind of flow performed. When `False` (the default),\n",
      "        this function attempts to open the user's web browser before falling back to\n",
      "        an out-of-band flow. When `True`, the out-of-band flow is always used.\n",
      "        \"\"\"\n",
      "\n",
      "        # This function and the components that it relies on are based off of:\n",
      "        # https://github.com/psteniusubi/python-sample\n",
      "\n",
      "        from sigstore._internal.oidc.oauth import _OAuthFlow\n",
      "\n",
      "        code: str\n",
      "        with _OAuthFlow(client_id, client_secret, self) as server:\n",
      "            # Launch web browser\n",
      "            if not force_oob and webbrowser.open(server.base_uri):\n",
      "                print(\"Waiting for browser interaction...\", file=sys.stderr)\n",
      "            else:\n",
      "                server.enable_oob()\n",
      "                print(\n",
      "                    f\"Go to the following link in a browser:\\n\\n\\t{server.auth_endpoint}\",\n",
      "                    file=sys.stderr,\n",
      "                )\n",
      "\n",
      "            if not server.is_oob():\n",
      "                # Wait until the redirect server populates the response\n",
      "                while server.auth_response is None:\n",
      "                    time.sleep(0.1)\n",
      "\n",
      "                auth_error = server.auth_response.get(\"error\")\n",
      "                if auth_error is not None:\n",
      "                    raise IdentityError(\n",
      "                        f\"Error response from auth endpoint: {auth_error[0]}\"\n",
      "                    )\n",
      "                code = server.auth_response[\"code\"][0]\n",
      "            else:\n",
      "                # In the out-of-band case, we wait until the user provides the code\n",
      "                code = input(\"Enter verification code: \")\n",
      "\n",
      "        # Provide code to token endpoint\n",
      "        data = {\n",
      "            \"grant_type\": \"authorization_code\",\n",
      "            \"redirect_uri\": server.redirect_uri,\n",
      "            \"code\": code,\n",
      "            \"code_verifier\": server.oauth_session.code_verifier,\n",
      "        }\n",
      "        auth = (\n",
      "            client_id,\n",
      "            client_secret,\n",
      "        )\n",
      "        logging.debug(f\"PAYLOAD: data={data}\")\n",
      "        try:\n",
      "            resp: requests.Response = requests.post(\n",
      "                self.oidc_config.token_endpoint,\n",
      "                data=data,\n",
      "                auth=auth,\n",
      "                timeout=30,\n",
      "            )\n",
      "        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "            raise NetworkError from exc\n",
      "\n",
      "        try:\n",
      "            resp.raise_for_status()\n",
      "        except requests.HTTPError as http_error:\n",
      "            raise IdentityError(\n",
      "                f\"Token request failed with {resp.status_code}\"\n",
      "            ) from http_error\n",
      "\n",
      "        token_json = resp.json()\n",
      "        token_error = token_json.get(\"error\")\n",
      "        if token_error is not None:\n",
      "            raise IdentityError(f\"Error response from token endpoint: {token_error}\")\n",
      "\n",
      "        return IdentityToken(token_json[\"access_token\"])\n",
      "\n",
      "\n",
      "class IdentityError(Error):\n",
      "    \"\"\"\n",
      "    Wraps `id`'s IdentityError.\n",
      "    \"\"\"\n",
      "\n",
      "    @classmethod\n",
      "    def raise_from_id(cls, exc: id.IdentityError) -> NoReturn:\n",
      "        \"\"\"Raises a wrapped IdentityError from the provided `id.IdentityError`.\"\"\"\n",
      "        raise cls(str(exc)) from exc\n",
      "\n",
      "    def diagnostics(self) -> str:\n",
      "        \"\"\"Returns diagnostics for the error.\"\"\"\n",
      "        if isinstance(self.__cause__, id.GitHubOidcPermissionCredentialError):\n",
      "            return f\"\"\"\n",
      "                Insufficient permissions for GitHub Actions workflow.\n",
      "\n",
      "                The most common reason for this is incorrect\n",
      "                configuration of the top-level `permissions` setting of the\n",
      "                workflow YAML file. It should be configured like so:\n",
      "\n",
      "                    permissions:\n",
      "                      id-token: write\n",
      "\n",
      "                Relevant documentation here:\n",
      "\n",
      "                    https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#adding-permissions-settings\n",
      "\n",
      "                Another possible reason is that the workflow run has been\n",
      "                triggered by a PR from a forked repository. PRs from forked\n",
      "                repositories typically cannot be granted write access.\n",
      "\n",
      "                Relevant documentation here:\n",
      "\n",
      "                    https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token\n",
      "\n",
      "                Additional context:\n",
      "\n",
      "                {self.__cause__}\n",
      "                \"\"\"\n",
      "        else:\n",
      "            return f\"\"\"\n",
      "                An issue occurred with ambient credential detection.\n",
      "\n",
      "                Additional context:\n",
      "\n",
      "                {self}\n",
      "            \"\"\"\n",
      "\n",
      "\n",
      "def detect_credential() -> Optional[str]:\n",
      "    \"\"\"Calls `id.detect_credential`, but wraps exceptions with our own exception type.\"\"\"\n",
      "    try:\n",
      "        return cast(Optional[str], id.detect_credential(_DEFAULT_AUDIENCE))\n",
      "    except id.IdentityError as exc:\n",
      "        IdentityError.raise_from_id(exc)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(code_snippet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "914a6ca6-0f6c-4156-ab8d-06f9586cb205",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.782372Z",
     "iopub.status.busy": "2023-11-21T15:30:47.782044Z",
     "iopub.status.idle": "2023-11-21T15:30:47.785930Z",
     "shell.execute_reply": "2023-11-21T15:30:47.785421Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.782354Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15939"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(code_snippet)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2966add0-f71b-4ea8-a655-e2dbaf3f8a77",
   "metadata": {},
   "source": [
    "Since the length of the code is big, let's try the prompts on a smaller code piece."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "936874ef-5421-4b35-b399-3d1caceb2710",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.786984Z",
     "iopub.status.busy": "2023-11-21T15:30:47.786637Z",
     "iopub.status.idle": "2023-11-21T15:30:47.791243Z",
     "shell.execute_reply": "2023-11-21T15:30:47.790722Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.786965Z"
    }
   },
   "outputs": [],
   "source": [
    "def split_code_into_pieces(code):\n",
    "    # Use regular expressions to split based on patterns (e.g., functions, classes)\n",
    "    pattern = r'\\b(?:def|class)\\s+[a-zA-Z_][a-zA-Z0-9_]*\\s*\\([^)]*\\)\\s*:'\n",
    "    code_pieces = re.split(pattern, code)\n",
    "    \n",
    "    # Include the matched patterns as separators between code pieces\n",
    "    separators = re.findall(pattern, code)\n",
    "    \n",
    "    # Combine the code pieces and separators into a list\n",
    "    code_sections = [piece.strip() for piece in code_pieces if piece.strip()]\n",
    "    \n",
    "    # Combine code pieces with separators in between\n",
    "    result_code = []\n",
    "    for i, section in enumerate(code_sections):\n",
    "        result_code.append(section)\n",
    "        if i < len(separators):\n",
    "            result_code.append(separators[i])\n",
    "\n",
    "    return result_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e59496d4-5271-4e86-bea9-bfe7def9e089",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.792268Z",
     "iopub.status.busy": "2023-11-21T15:30:47.791950Z",
     "iopub.status.idle": "2023-11-21T15:30:47.797331Z",
     "shell.execute_reply": "2023-11-21T15:30:47.796812Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.792250Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "code_pieces = split_code_into_pieces(code_snippet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fbd9ef9e-6d04-4f7a-b3bb-476c91bc755b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.798388Z",
     "iopub.status.busy": "2023-11-21T15:30:47.798058Z",
     "iopub.status.idle": "2023-11-21T15:30:47.801583Z",
     "shell.execute_reply": "2023-11-21T15:30:47.801064Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.798370Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"\"\"\n",
      "    Raised on any communication or format error with an OIDC issuer.\n",
      "    \"\"\"\n",
      "\n",
      "    pass\n",
      "\n",
      "\n",
      "class Issuer:\n",
      "    \"\"\"\n",
      "    Represents an OIDC issuer (IdP).\n",
      "    \"\"\"\n",
      "\n",
      "    def __init__(self, base_url: str) -> None:\n",
      "        \"\"\"\n",
      "        Create a new `Issuer` from the given base URL.\n",
      "\n",
      "        This URL is used to locate an OpenID Connect configuration file,\n",
      "        which is then used to bootstrap the issuer's state (such\n",
      "        as authorization and token endpoints).\n",
      "        \"\"\"\n",
      "        oidc_config_url = urllib.parse.urljoin(\n",
      "            f\"{base_url}/\", \".well-known/openid-configuration\"\n",
      "        )\n",
      "\n",
      "        try:\n",
      "            resp: requests.Response = requests.get(oidc_config_url, timeout=30)\n",
      "        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "            raise NetworkError from exc\n",
      "\n",
      "        try:\n",
      "            resp.raise_for_status()\n",
      "        except requests.HTTPError as http_error:\n",
      "            raise IssuerError from http_error\n",
      "\n",
      "        try:\n",
      "            # We don't generally expect this to fail (since the provider should\n",
      "            # return a non-success HTTP code which we catch above), but we\n",
      "            # check just in case we have a misbehaving OIDC issuer.\n",
      "            self.oidc_config = _OpenIDConfiguration.model_validate(resp.json())\n",
      "        except ValueError as exc:\n",
      "            raise IssuerError(f\"OIDC issuer returned invalid configuration: {exc}\")\n",
      "\n",
      "    @classmethod\n",
      "    def production(cls) -> Issuer:\n",
      "        \"\"\"\n",
      "        Returns an `Issuer` configured against Sigstore's production-level services.\n",
      "        \"\"\"\n",
      "        return cls(DEFAULT_OAUTH_ISSUER_URL)\n",
      "\n",
      "    @classmethod\n",
      "    def staging(cls) -> Issuer:\n",
      "        \"\"\"\n",
      "        Returns an `Issuer` configured against Sigstore's staging-level services.\n",
      "        \"\"\"\n",
      "        return cls(STAGING_OAUTH_ISSUER_URL)\n",
      "\n",
      "    def identity_token(  # nosec: B107\n",
      "        self,\n",
      "        client_id: str = \"sigstore\",\n",
      "        client_secret: str = \"\",\n",
      "        force_oob: bool = False,\n",
      "    ) -> IdentityToken:\n",
      "        \"\"\"\n",
      "        Retrieves and returns an `IdentityToken` from the current `Issuer`, via OAuth.\n",
      "\n",
      "        This function blocks on user interaction.\n",
      "\n",
      "        The `force_oob` flag controls the kind of flow performed. When `False` (the default),\n",
      "        this function attempts to open the user's web browser before falling back to\n",
      "        an out-of-band flow. When `True`, the out-of-band flow is always used.\n",
      "        \"\"\"\n",
      "\n",
      "        # This function and the components that it relies on are based off of:\n",
      "        # https://github.com/psteniusubi/python-sample\n",
      "\n",
      "        from sigstore._internal.oidc.oauth import _OAuthFlow\n",
      "\n",
      "        code: str\n",
      "        with _OAuthFlow(client_id, client_secret, self) as server:\n",
      "            # Launch web browser\n",
      "            if not force_oob and webbrowser.open(server.base_uri):\n",
      "                print(\"Waiting for browser interaction...\", file=sys.stderr)\n",
      "            else:\n",
      "                server.enable_oob()\n",
      "                print(\n",
      "                    f\"Go to the following link in a browser:\\n\\n\\t{server.auth_endpoint}\",\n",
      "                    file=sys.stderr,\n",
      "                )\n",
      "\n",
      "            if not server.is_oob():\n",
      "                # Wait until the redirect server populates the response\n",
      "                while server.auth_response is None:\n",
      "                    time.sleep(0.1)\n",
      "\n",
      "                auth_error = server.auth_response.get(\"error\")\n",
      "                if auth_error is not None:\n",
      "                    raise IdentityError(\n",
      "                        f\"Error response from auth endpoint: {auth_error[0]}\"\n",
      "                    )\n",
      "                code = server.auth_response[\"code\"][0]\n",
      "            else:\n",
      "                # In the out-of-band case, we wait until the user provides the code\n",
      "                code = input(\"Enter verification code: \")\n",
      "\n",
      "        # Provide code to token endpoint\n",
      "        data = {\n",
      "            \"grant_type\": \"authorization_code\",\n",
      "            \"redirect_uri\": server.redirect_uri,\n",
      "            \"code\": code,\n",
      "            \"code_verifier\": server.oauth_session.code_verifier,\n",
      "        }\n",
      "        auth = (\n",
      "            client_id,\n",
      "            client_secret,\n",
      "        )\n",
      "        logging.debug(f\"PAYLOAD: data={data}\")\n",
      "        try:\n",
      "            resp: requests.Response = requests.post(\n",
      "                self.oidc_config.token_endpoint,\n",
      "                data=data,\n",
      "                auth=auth,\n",
      "                timeout=30,\n",
      "            )\n",
      "        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "            raise NetworkError from exc\n",
      "\n",
      "        try:\n",
      "            resp.raise_for_status()\n",
      "        except requests.HTTPError as http_error:\n",
      "            raise IdentityError(\n",
      "                f\"Token request failed with {resp.status_code}\"\n",
      "            ) from http_error\n",
      "\n",
      "        token_json = resp.json()\n",
      "        token_error = token_json.get(\"error\")\n",
      "        if token_error is not None:\n",
      "            raise IdentityError(f\"Error response from token endpoint: {token_error}\")\n",
      "\n",
      "        return IdentityToken(token_json[\"access_token\"])\n"
     ]
    }
   ],
   "source": [
    "print(code_pieces[6])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02d94eaf-42c1-47a3-b536-0c42c4e4b80f",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Generate prompts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "940cecfa-69a3-42b3-87b6-af796c0c3cfb",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.802650Z",
     "iopub.status.busy": "2023-11-21T15:30:47.802311Z",
     "iopub.status.idle": "2023-11-21T15:30:47.805620Z",
     "shell.execute_reply": "2023-11-21T15:30:47.805114Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.802631Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def generate_text(chosen_model, prompt):\n",
    "    response = chosen_model.generate([prompt_1])\n",
    "    generated_text = response[0].generated_text\n",
    "    return generated_text"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b58cb2c3-648c-4d06-a120-78b52ab60d67",
   "metadata": {},
   "source": [
    "#### Prompt 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4b4cf88d-3c67-4f18-beba-d6c70f015cff",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.806692Z",
     "iopub.status.busy": "2023-11-21T15:30:47.806336Z",
     "iopub.status.idle": "2023-11-21T15:30:47.809703Z",
     "shell.execute_reply": "2023-11-21T15:30:47.809194Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.806673Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Example prompt 1\n",
    "prompt_1 = \"\"\"\n",
    "Generate documentation for each function in the given \n",
    "Code snippet:{code_snippet}\n",
    "\n",
    "Function or class name\n",
    "\n",
    "\"\"\".format(code_snippet=code_pieces[6])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5bbb85f1-940c-4f08-bf63-7eb67fb66c7c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:47.810761Z",
     "iopub.status.busy": "2023-11-21T15:30:47.810416Z",
     "iopub.status.idle": "2023-11-21T15:30:58.976933Z",
     "shell.execute_reply": "2023-11-21T15:30:58.976284Z",
     "shell.execute_reply.started": "2023-11-21T15:30:47.810742Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"\"\"\n",
      "    def id_token(  # nosec: B107\n",
      "        self,\n",
      "        force_oob: bool = False,\n",
      "    ) -> IdentityToken:\n",
      "        \"\"\"\n",
      "        Retrieves and returns an `IdentityToken` from the current `Issuer`, via OAuth.\n",
      "\n",
      "        This function blocks on user interaction.\n",
      "\n",
      "        The `force_oob` flag controls the kind of flow performed. When `False` (the default),\n",
      "        this function attempts to open the user's web browser before falling back to\n",
      "        an out-of-band flow. When `True`, the out-of-band flow is always used.\n",
      "        \"\"\"\n",
      "\n",
      "        # This function and the components that it relies on are based off of:\n",
      "        # https://github.com/psteniusubi/python-sample\n",
      "\n",
      "        from sigstore._internal.oidc.oauth import _OAuthFlow\n",
      "\n",
      "        code: str\n",
      "        with _OAuthFlow(self) as server:\n",
      "            # Launch web browser\n",
      "            if not force_oob and webbrowser.open(server.base_uri):\n",
      "                print(\"Waiting for browser interaction...\", file=sys.stderr)\n",
      "            else:\n",
      "                server.enable_oob()\n",
      "                print(\n",
      "                    f\"Go to the following link in a browser:\\n\\n\\t{server.auth_endpoint}\",\n",
      "                    file=sys.stderr,\n",
      "                )\n",
      "\n",
      "            if not server.is_oob():\n",
      "                # Wait until the redirect server populates the response\n",
      "                while server.auth_response is None:\n",
      "                    time.sleep(0.1)\n",
      "\n",
      "                auth_error = server.auth_response.get(\"error\")\n",
      "                if auth_error is not None:\n",
      "                    raise IdentityError(\n",
      "                        f\"Error response from auth endpoint: {auth_error[0]}\"\n",
      "                    )\n",
      "                code = server.auth_response[\"code\"][0]\n",
      "            else:\n",
      "                # In the out-of-band case, we wait until the user provides the code\n",
      "                code = input(\"Enter verification code: \")\n",
      "\n",
      "        # Provide code to token endpoint\n",
      "        data = {\n",
      "            \"grant_type\": \"authorization_code\",\n",
      "            \"redirect_uri\": server.redirect_uri,\n",
      "            \"code\": code,\n",
      "            \"code_verifier\": server.oauth_session.code_verifier,\n",
      "        }\n",
      "        auth = (\n",
      "            self.client_id,\n",
      "            self.client_secret,\n",
      "        )\n",
      "        logging.debug(f\"PAYLOAD: data={data}\")\n",
      "        try:\n",
      "            resp: requests.Response = requests.post(\n",
      "                self.oidc_config.token_endpoint,\n",
      "                data=data,\n",
      "                auth=auth,\n",
      "                timeout=30,\n",
      "            )\n",
      "        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "            raise NetworkError from exc\n",
      "\n",
      "        try:\n",
      "            resp.raise_for_status()\n",
      "        except requests.HTTPError as http_error:\n",
      "            raise IdentityError(\n",
      "                f\"Token request failed with {resp.status_code}\"\n",
      "            ) from http_error\n",
      "\n",
      "        token_json = resp.json()\n",
      "        token_error = token_json.get(\"error\")\n",
      "        if token_error is not None:\n",
      "            raise IdentityError(f\"Error response from token endpoint: {token_error}\")\n",
      "\n",
      "        return IdentityToken(token_json[\"access_token\"])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(generate_text(chosen_model, prompt_1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dbb80f5-09db-44ed-a2d4-30155f285582",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Prompt 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "302edb86-3df8-4eac-a35f-a68212d57371",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:58.978112Z",
     "iopub.status.busy": "2023-11-21T15:30:58.977799Z",
     "iopub.status.idle": "2023-11-21T15:30:58.981324Z",
     "shell.execute_reply": "2023-11-21T15:30:58.980743Z",
     "shell.execute_reply.started": "2023-11-21T15:30:58.978090Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Example prompt 2\n",
    "prompt_2 = \"\"\"\n",
    "Generate documentation for each function in the given code snippet.\n",
    "\n",
    "1. Function name \n",
    "2. Function description \n",
    "3. Input arguments\n",
    "4. Parameters\n",
    "\n",
    "Code snippet:\n",
    "{code_snippet}\n",
    "\"\"\".format(code_snippet=code_pieces[6])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "92c4d4e1-f16e-4910-953f-d917b14dca63",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:30:58.982337Z",
     "iopub.status.busy": "2023-11-21T15:30:58.982083Z",
     "iopub.status.idle": "2023-11-21T15:31:13.087942Z",
     "shell.execute_reply": "2023-11-21T15:31:13.087292Z",
     "shell.execute_reply.started": "2023-11-21T15:30:58.982318Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    def __init__(self, base_url: str) -> None:\n",
      "        \"\"\"\n",
      "        Create a new `Issuer` from the given base URL.\n",
      "\n",
      "        This URL is used to locate an OpenID Connect configuration file,\n",
      "        which is then used to bootstrap the issuer's state (such\n",
      "        as authorization and token endpoints).\n",
      "        \"\"\"\n",
      "        oidc_config_url = urllib.parse.urljoin(\n",
      "            f\"{base_url}/\", \".well-known/openid-configuration\"\n",
      "        )\n",
      "\n",
      "        try:\n",
      "            resp: requests.Response = requests.get(oidc_config_url, timeout=30)\n",
      "        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "            raise NetworkError from exc\n",
      "\n",
      "        try:\n",
      "            resp.raise_for_status()\n",
      "        except requests.HTTPError as http_error:\n",
      "            raise IssuerError from http_error\n",
      "\n",
      "        try:\n",
      "            # We don't generally expect this to fail (since the provider should\n",
      "            # return a non-success HTTP code which we catch above), but we\n",
      "            # check just in case we have a misbehaving OIDC issuer.\n",
      "            self.oidc_config = _OpenIDConfiguration.model_validate(resp.json())\n",
      "        except ValueError as exc:\n",
      "            raise IssuerError(f\"OIDC issuer returned invalid configuration: {exc}\")\n",
      "\n",
      "    @classmethod\n",
      "    def production(cls) -> Issuer:\n",
      "        \"\"\"\n",
      "        Returns an `Issuer` configured against Sigstore's production-level services.\n",
      "        \"\"\"\n",
      "        return cls(DEFAULT_OAUTH_ISSUER_URL)\n",
      "\n",
      "    @classmethod\n",
      "    def staging(cls) -> Issuer:\n",
      "        \"\"\"\n",
      "        Returns an `Issuer` configured against Sigstore's staging-level services.\n",
      "        \"\"\"\n",
      "        return cls(STAGING_OAUTH_ISSUER_URL)\n",
      "\n",
      "    def identity_token(  # nosec: B107\n",
      "        self,\n",
      "        client_id: str = \"sigstore\",\n",
      "        client_secret: str = \"\",\n",
      "        force_oob: bool = False,\n",
      "    ) -> IdentityToken:\n",
      "        \"\"\"\n",
      "        Retrieves and returns an `IdentityToken` from the current `Issuer`, via OAuth.\n",
      "\n",
      "        This function blocks on user interaction.\n",
      "\n",
      "        The `force_oob` flag controls the kind of flow performed. When `False` (the default),\n",
      "        this function attempts to open the user's web browser before falling back to\n",
      "        an out-of-band flow. When `True`, the out-of-band flow is always used.\n",
      "        \"\"\"\n",
      "\n",
      "        # This function and the components that it relies on are based off of:\n",
      "        # https://github.com/psteniusubi/python-sample\n",
      "\n",
      "        from sigstore._internal.oidc.oauth import _OAuthFlow\n",
      "\n",
      "        code: str\n",
      "        with _OAuthFlow(client_id, client_secret, self) as server:\n",
      "            # Launch web browser\n",
      "            if not force_oob and webbrowser.open(server.base_uri):\n",
      "                print(\"Waiting for browser interaction...\", file=sys.stderr)\n",
      "            else:\n",
      "                server.enable_oob()\n",
      "                print(\n",
      "                    f\"Go to the following link in a browser:\\n\\n\\t{server.auth_endpoint}\",\n",
      "                    file=sys.stderr,\n",
      "                )\n",
      "\n",
      "            if not server.is_oob():\n",
      "                # Wait until the redirect server populates the response\n",
      "                while server.auth_response is None:\n",
      "                    time.sleep(0.1)\n",
      "\n",
      "                auth_error = server.auth_response.get(\"error\")\n",
      "                if auth_error is not None:\n",
      "                    raise IdentityError(\n",
      "                        f\"Error response from auth endpoint: {auth_error[0]}\"\n",
      "                    )\n",
      "                code = server.auth_response[\"code\"][0]\n",
      "            else:\n",
      "                # In the out-of-band case, we wait until the user provides the code\n",
      "                code = input(\"Enter verification code: \")\n",
      "\n",
      "        # Provide code to token endpoint\n",
      "        data = {\n",
      "            \"grant_type\": \"authorization_code\",\n",
      "            \"redirect_uri\": server.redirect_uri,\n",
      "            \"code\": code,\n",
      "            \"code_verifier\": server.oauth_session.code_verifier,\n",
      "        }\n"
     ]
    }
   ],
   "source": [
    "print(generate_text(chosen_model, prompt_2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d76eb787-8140-4598-8b3b-61b3ae903cb9",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Expected Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b4149f0b-c6b5-43cd-a0c8-7ae08d4f0b6b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T15:31:13.089153Z",
     "iopub.status.busy": "2023-11-21T15:31:13.088776Z",
     "iopub.status.idle": "2023-11-21T15:31:13.093253Z",
     "shell.execute_reply": "2023-11-21T15:31:13.092721Z",
     "shell.execute_reply.started": "2023-11-21T15:31:13.089132Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ sigstore](../sigstore.html)\n",
      "\n",
      "## API Documentation\n",
      "\n",
      "  * DEFAULT_OAUTH_ISSUER_URL\n",
      "  * STAGING_OAUTH_ISSUER_URL\n",
      "  * DEFAULT_AUDIENCE\n",
      "  * ExpiredIdentity\n",
      "  * IdentityToken\n",
      "    * IdentityToken\n",
      "    * in_validity_period\n",
      "    * identity\n",
      "    * issuer\n",
      "    * expected_certificate_subject\n",
      "  * IssuerError\n",
      "  * Issuer\n",
      "    * Issuer\n",
      "    * production\n",
      "    * staging\n",
      "    * identity_token\n",
      "  * IdentityError\n",
      "    * raise_from_id\n",
      "    * diagnostics\n",
      "  * detect_credential\n",
      "\n",
      "[ built with pdoc ](https://pdoc.dev \"pdoc: Python API documentation\n",
      "generator\")\n",
      "\n",
      "#  [sigstore](./../sigstore.html).oidc\n",
      "\n",
      "API for retrieving OIDC tokens.\n",
      "\n",
      "View Source\n",
      "\n",
      "    \n",
      "    \n",
      "      1# Copyright 2022 The Sigstore Authors\n",
      "      2#\n",
      "      3# Licensed under the Apache License, Version 2.0 (the \"License\");\n",
      "      4# you may not use this file except in compliance with the License.\n",
      "      5# You may obtain a copy of the License at\n",
      "      6#\n",
      "      7#      http://www.apache.org/licenses/LICENSE-2.0\n",
      "      8#\n",
      "      9# Unless required by applicable law or agreed to in writing, software\n",
      "     10# distributed under the License is distributed on an \"AS IS\" BASIS,\n",
      "     11# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
      "     12# See the License for the specific language governing permissions and\n",
      "     13# limitations under the License.\n",
      "     14\n",
      "     15\"\"\"\n",
      "     16API for retrieving OIDC tokens.\n",
      "     17\"\"\"\n",
      "     18\n",
      "     19from __future__ import annotations\n",
      "     20\n",
      "     21import logging\n",
      "     22import sys\n",
      "     23import time\n",
      "     24import urllib.parse\n",
      "     25import webbrowser\n",
      "     26from datetime import datetime, timezone\n",
      "     27from typing import NoReturn, Optional, cast\n",
      "     28\n",
      "     29import id\n",
      "     30import jwt\n",
      "     31import requests\n",
      "     32from pydantic import BaseModel, StrictStr\n",
      "     33\n",
      "     34from sigstore.errors import Error, NetworkError\n",
      "     35\n",
      "     36DEFAULT_OAUTH_ISSUER_URL = \"https://oauth2.sigstore.dev/auth\"\n",
      "     37STAGING_OAUTH_ISSUER_URL = \"https://oauth2.sigstage.dev/auth\"\n",
      "     38\n",
      "     39# See: https://github.com/sigstore/fulcio/blob/b2186c0/pkg/config/config.go#L182-L201\n",
      "     40_KNOWN_OIDC_ISSUERS = {\n",
      "     41    \"https://accounts.google.com\": \"email\",\n",
      "     42    \"https://oauth2.sigstore.dev/auth\": \"email\",\n",
      "     43    \"https://oauth2.sigstage.dev/auth\": \"email\",\n",
      "     44    \"https://token.actions.githubusercontent.com\": \"sub\",\n",
      "     45}\n",
      "     46_DEFAULT_AUDIENCE = \"sigstore\"\n",
      "     47\n",
      "     48\n",
      "     49class _OpenIDConfiguration(BaseModel):\n",
      "     50    \"\"\"\n",
      "     51    Represents a (subset) of the fields provided by an OpenID Connect provider's\n",
      "     52    `.well-known/openid-configuration` response, as defined by OpenID Connect Discovery.\n",
      "     53\n",
      "     54    See: <https://openid.net/specs/openid-connect-discovery-1_0.html>\n",
      "     55    \"\"\"\n",
      "     56\n",
      "     57    authorization_endpoint: StrictStr\n",
      "     58    token_endpoint: StrictStr\n",
      "     59\n",
      "     60\n",
      "     61# See: https://github.com/sigstore/fulcio/blob/b2186c0/pkg/config/config.go#L182-L201\n",
      "     62_KNOWN_OIDC_ISSUERS = {\n",
      "     63    \"https://accounts.google.com\": \"email\",\n",
      "     64    \"https://oauth2.sigstore.dev/auth\": \"email\",\n",
      "     65    \"https://oauth2.sigstage.dev/auth\": \"email\",\n",
      "     66    \"https://token.actions.githubusercontent.com\": \"sub\",\n",
      "     67}\n",
      "     68DEFAULT_AUDIENCE = \"sigstore\"\n",
      "     69\n",
      "     70\n",
      "     71class ExpiredIdentity(Exception):\n",
      "     72    \"\"\"An error raised when an identity token is expired.\"\"\"\n",
      "     73\n",
      "     74\n",
      "     75class IdentityToken:\n",
      "     76    \"\"\"\n",
      "     77    An OIDC \"identity\", corresponding to an underlying OIDC token with\n",
      "     78    a sensible subject, issuer, and audience for Sigstore purposes.\n",
      "     79    \"\"\"\n",
      "     80\n",
      "     81    def __init__(self, raw_token: str) -> None:\n",
      "     82        \"\"\"\n",
      "     83        Create a new `IdentityToken` from the given OIDC token.\n",
      "     84        \"\"\"\n",
      "     85\n",
      "     86        self._raw_token = raw_token\n",
      "     87\n",
      "     88        # NOTE: The lack of verification here is intentional, and is part of\n",
      "     89        # Sigstore's verification model: clients like sigstore-python are\n",
      "     90        # responsible only for forwarding the OIDC identity to Fulcio for\n",
      "     91        # certificate binding and issuance.\n",
      "     92        try:\n",
      "     93            self._unverified_claims = jwt.decode(\n",
      "     94                raw_token,\n",
      "     95                options={\n",
      "     96                    \"verify_signature\": False,\n",
      "     97                    \"verify_aud\": True,\n",
      "     98                    \"verify_iat\": True,\n",
      "     99                    \"verify_exp\": True,\n",
      "    100                    # These claims are required by OpenID Connect, so\n",
      "    101                    # we can strongly enforce their presence.\n",
      "    102                    # See: https://openid.net/specs/openid-connect-basic-1_0.html#IDToken\n",
      "    103                    \"require\": [\"aud\", \"sub\", \"iat\", \"exp\", \"iss\"],\n",
      "    104                },\n",
      "    105                audience=DEFAULT_AUDIENCE,\n",
      "    106                # NOTE: This leeway shouldn't be strictly necessary, but is\n",
      "    107                # included to preempt any (small) skew between the host\n",
      "    108                # and the originating IdP.\n",
      "    109                leeway=5,\n",
      "    110            )\n",
      "    111        except Exception as exc:\n",
      "    112            raise IdentityError(\n",
      "    113                \"Identity token is malformed or missing claims\"\n",
      "    114            ) from exc\n",
      "    115\n",
      "    116        self._iss: str = self._unverified_claims[\"iss\"]\n",
      "    117        self._nbf: int | None = self._unverified_claims.get(\"nbf\")\n",
      "    118        self._exp: int = self._unverified_claims[\"exp\"]\n",
      "    119\n",
      "    120        # Fail early if this token isn't within its validity period.\n",
      "    121        if not self.in_validity_period():\n",
      "    122            raise IdentityError(\"Identity token is not within its validity period\")\n",
      "    123\n",
      "    124        # When verifying the private key possession proof, Fulcio uses\n",
      "    125        # different claims depending on the token's issuer.\n",
      "    126        # We currently special-case a handful of these, and fall back\n",
      "    127        # on signing the \"sub\" claim otherwise.\n",
      "    128        identity_claim = _KNOWN_OIDC_ISSUERS.get(self.issuer)\n",
      "    129        if identity_claim is not None:\n",
      "    130            if identity_claim not in self._unverified_claims:\n",
      "    131                raise IdentityError(\n",
      "    132                    f\"Identity token is missing the required {identity_claim!r} claim\"\n",
      "    133                )\n",
      "    134\n",
      "    135            self._identity = str(self._unverified_claims.get(identity_claim))\n",
      "    136        else:\n",
      "    137            try:\n",
      "    138                self._identity = str(self._unverified_claims[\"sub\"])\n",
      "    139            except KeyError:\n",
      "    140                raise IdentityError(\n",
      "    141                    \"Identity token is missing the required 'sub' claim\"\n",
      "    142                )\n",
      "    143\n",
      "    144        # This identity token might have been retrieved directly from\n",
      "    145        # an identity provider, or it might be a \"federated\" identity token\n",
      "    146        # retrieved from a federated IdP (e.g., Sigstore's own Dex instance).\n",
      "    147        # In the latter case, the claims will also include a `federated_claims`\n",
      "    148        # set, which in turn should include a `connector_id` that reflects\n",
      "    149        # the \"real\" token issuer. We retrieve this, despite technically\n",
      "    150        # being an implementation detail, because it has value to client\n",
      "    151        # users: a client might want to make sure that its user is identifying\n",
      "    152        # with a *particular* IdP, which means that they need to pierce the\n",
      "    153        # federation layer to check which IdP is actually being used.\n",
      "    154        self._federated_issuer: str | None = None\n",
      "    155        federated_claims = self._unverified_claims.get(\"federated_claims\")\n",
      "    156        if federated_claims is not None:\n",
      "    157            if not isinstance(federated_claims, dict):\n",
      "    158                raise IdentityError(\n",
      "    159                    \"unexpected claim type: federated_claims is not a dict\"\n",
      "    160                )\n",
      "    161\n",
      "    162            federated_issuer = federated_claims.get(\"connector_id\")\n",
      "    163            if federated_issuer is not None:\n",
      "    164                if not isinstance(federated_issuer, str):\n",
      "    165                    raise IdentityError(\n",
      "    166                        \"unexpected claim type: federated_claims.connector_id is not a string\"\n",
      "    167                    )\n",
      "    168\n",
      "    169                self._federated_issuer = federated_issuer\n",
      "    170\n",
      "    171    def in_validity_period(self) -> bool:\n",
      "    172        \"\"\"\n",
      "    173        Returns whether or not this `Identity` is currently within its self-stated validity period.\n",
      "    174\n",
      "    175        NOTE: As noted in `Identity.__init__`, this is not a verifying wrapper;\n",
      "    176        the check here only asserts whether the *unverified* identity's claims\n",
      "    177        are within their validity period.\n",
      "    178        \"\"\"\n",
      "    179\n",
      "    180        now = datetime.now(timezone.utc).timestamp()\n",
      "    181\n",
      "    182        if self._nbf is not None:\n",
      "    183            return self._nbf <= now < self._exp\n",
      "    184        else:\n",
      "    185            return now < self._exp\n",
      "    186\n",
      "    187    @property\n",
      "    188    def identity(self) -> str:\n",
      "    189        \"\"\"\n",
      "    190        Returns this `IdentityToken`'s underlying \"subject\".\n",
      "    191\n",
      "    192        Note that this is **not** always the `sub` claim in the corresponding\n",
      "    193        identity token: depending onm the token's issuer, it may be a *different*\n",
      "    194        claim, such as `email`. This corresponds to the Sigstore ecosystem's\n",
      "    195        behavior, e.g. in each issued certificate's SAN.\n",
      "    196        \"\"\"\n",
      "    197        return self._identity\n",
      "    198\n",
      "    199    @property\n",
      "    200    def issuer(self) -> str:\n",
      "    201        \"\"\"\n",
      "    202        Returns a URL identifying this `IdentityToken`'s issuer.\n",
      "    203        \"\"\"\n",
      "    204        return self._iss\n",
      "    205\n",
      "    206    @property\n",
      "    207    def expected_certificate_subject(self) -> str:\n",
      "    208        \"\"\"\n",
      "    209        Returns a URL identifying the **expected** subject for any Sigstore\n",
      "    210        certificate issued against this identity token.\n",
      "    211\n",
      "    212        The behavior of this field is slightly subtle: for non-federated\n",
      "    213        identity providers (like a token issued directly by Google's IdP) it\n",
      "    214        should be exactly equivalent to `IdentityToken.issuer`. For federated\n",
      "    215        issuers (like Sigstore's own federated IdP) it should be equivalent to\n",
      "    216        the underlying federated issuer's URL, which is kept in an\n",
      "    217        implementation-defined claim.\n",
      "    218\n",
      "    219        This attribute exists so that clients who wish to inspect the expected\n",
      "    220        subject of their certificates can do so without relying on\n",
      "    221        implementation-specific behavior.\n",
      "    222        \"\"\"\n",
      "    223        if self._federated_issuer is not None:\n",
      "    224            return self._federated_issuer\n",
      "    225\n",
      "    226        return self.issuer\n",
      "    227\n",
      "    228    def __str__(self) -> str:\n",
      "    229        \"\"\"\n",
      "    230        Returns the underlying OIDC token for this identity.\n",
      "    231\n",
      "    232        That this token is secret in nature and **MUST NOT** be disclosed.\n",
      "    233        \"\"\"\n",
      "    234        return self._raw_token\n",
      "    235\n",
      "    236\n",
      "    237class IssuerError(Exception):\n",
      "    238    \"\"\"\n",
      "    239    Raised on any communication or format error with an OIDC issuer.\n",
      "    240    \"\"\"\n",
      "    241\n",
      "    242    pass\n",
      "    243\n",
      "    244\n",
      "    245class Issuer:\n",
      "    246    \"\"\"\n",
      "    247    Represents an OIDC issuer (IdP).\n",
      "    248    \"\"\"\n",
      "    249\n",
      "    250    def __init__(self, base_url: str) -> None:\n",
      "    251        \"\"\"\n",
      "    252        Create a new `Issuer` from the given base URL.\n",
      "    253\n",
      "    254        This URL is used to locate an OpenID Connect configuration file,\n",
      "    255        which is then used to bootstrap the issuer's state (such\n",
      "    256        as authorization and token endpoints).\n",
      "    257        \"\"\"\n",
      "    258        oidc_config_url = urllib.parse.urljoin(\n",
      "    259            f\"{base_url}/\", \".well-known/openid-configuration\"\n",
      "    260        )\n",
      "    261\n",
      "    262        try:\n",
      "    263            resp: requests.Response = requests.get(oidc_config_url, timeout=30)\n",
      "    264        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "    265            raise NetworkError from exc\n",
      "    266\n",
      "    267        try:\n",
      "    268            resp.raise_for_status()\n",
      "    269        except requests.HTTPError as http_error:\n",
      "    270            raise IssuerError from http_error\n",
      "    271\n",
      "    272        try:\n",
      "    273            # We don't generally expect this to fail (since the provider should\n",
      "    274            # return a non-success HTTP code which we catch above), but we\n",
      "    275            # check just in case we have a misbehaving OIDC issuer.\n",
      "    276            self.oidc_config = _OpenIDConfiguration.model_validate(resp.json())\n",
      "    277        except ValueError as exc:\n",
      "    278            raise IssuerError(f\"OIDC issuer returned invalid configuration: {exc}\")\n",
      "    279\n",
      "    280    @classmethod\n",
      "    281    def production(cls) -> Issuer:\n",
      "    282        \"\"\"\n",
      "    283        Returns an `Issuer` configured against Sigstore's production-level services.\n",
      "    284        \"\"\"\n",
      "    285        return cls(DEFAULT_OAUTH_ISSUER_URL)\n",
      "    286\n",
      "    287    @classmethod\n",
      "    288    def staging(cls) -> Issuer:\n",
      "    289        \"\"\"\n",
      "    290        Returns an `Issuer` configured against Sigstore's staging-level services.\n",
      "    291        \"\"\"\n",
      "    292        return cls(STAGING_OAUTH_ISSUER_URL)\n",
      "    293\n",
      "    294    def identity_token(  # nosec: B107\n",
      "    295        self,\n",
      "    296        client_id: str = \"sigstore\",\n",
      "    297        client_secret: str = \"\",\n",
      "    298        force_oob: bool = False,\n",
      "    299    ) -> IdentityToken:\n",
      "    300        \"\"\"\n",
      "    301        Retrieves and returns an `IdentityToken` from the current `Issuer`, via OAuth.\n",
      "    302\n",
      "    303        This function blocks on user interaction.\n",
      "    304\n",
      "    305        The `force_oob` flag controls the kind of flow performed. When `False` (the default),\n",
      "    306        this function attempts to open the user's web browser before falling back to\n",
      "    307        an out-of-band flow. When `True`, the out-of-band flow is always used.\n",
      "    308        \"\"\"\n",
      "    309\n",
      "    310        # This function and the components that it relies on are based off of:\n",
      "    311        # https://github.com/psteniusubi/python-sample\n",
      "    312\n",
      "    313        from sigstore._internal.oidc.oauth import _OAuthFlow\n",
      "    314\n",
      "    315        code: str\n",
      "    316        with _OAuthFlow(client_id, client_secret, self) as server:\n",
      "    317            # Launch web browser\n",
      "    318            if not force_oob and webbrowser.open(server.base_uri):\n",
      "    319                print(\"Waiting for browser interaction...\", file=sys.stderr)\n",
      "    320            else:\n",
      "    321                server.enable_oob()\n",
      "    322                print(\n",
      "    323                    f\"Go to the following link in a browser:\\n\\n\\t{server.auth_endpoint}\",\n",
      "    324                    file=sys.stderr,\n",
      "    325                )\n",
      "    326\n",
      "    327            if not server.is_oob():\n",
      "    328                # Wait until the redirect server populates the response\n",
      "    329                while server.auth_response is None:\n",
      "    330                    time.sleep(0.1)\n",
      "    331\n",
      "    332                auth_error = server.auth_response.get(\"error\")\n",
      "    333                if auth_error is not None:\n",
      "    334                    raise IdentityError(\n",
      "    335                        f\"Error response from auth endpoint: {auth_error[0]}\"\n",
      "    336                    )\n",
      "    337                code = server.auth_response[\"code\"][0]\n",
      "    338            else:\n",
      "    339                # In the out-of-band case, we wait until the user provides the code\n",
      "    340                code = input(\"Enter verification code: \")\n",
      "    341\n",
      "    342        # Provide code to token endpoint\n",
      "    343        data = {\n",
      "    344            \"grant_type\": \"authorization_code\",\n",
      "    345            \"redirect_uri\": server.redirect_uri,\n",
      "    346            \"code\": code,\n",
      "    347            \"code_verifier\": server.oauth_session.code_verifier,\n",
      "    348        }\n",
      "    349        auth = (\n",
      "    350            client_id,\n",
      "    351            client_secret,\n",
      "    352        )\n",
      "    353        logging.debug(f\"PAYLOAD: data={data}\")\n",
      "    354        try:\n",
      "    355            resp: requests.Response = requests.post(\n",
      "    356                self.oidc_config.token_endpoint,\n",
      "    357                data=data,\n",
      "    358                auth=auth,\n",
      "    359                timeout=30,\n",
      "    360            )\n",
      "    361        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "    362            raise NetworkError from exc\n",
      "    363\n",
      "    364        try:\n",
      "    365            resp.raise_for_status()\n",
      "    366        except requests.HTTPError as http_error:\n",
      "    367            raise IdentityError(\n",
      "    368                f\"Token request failed with {resp.status_code}\"\n",
      "    369            ) from http_error\n",
      "    370\n",
      "    371        token_json = resp.json()\n",
      "    372        token_error = token_json.get(\"error\")\n",
      "    373        if token_error is not None:\n",
      "    374            raise IdentityError(f\"Error response from token endpoint: {token_error}\")\n",
      "    375\n",
      "    376        return IdentityToken(token_json[\"access_token\"])\n",
      "    377\n",
      "    378\n",
      "    379class IdentityError(Error):\n",
      "    380    \"\"\"\n",
      "    381    Wraps `id`'s IdentityError.\n",
      "    382    \"\"\"\n",
      "    383\n",
      "    384    @classmethod\n",
      "    385    def raise_from_id(cls, exc: id.IdentityError) -> NoReturn:\n",
      "    386        \"\"\"Raises a wrapped IdentityError from the provided `id.IdentityError`.\"\"\"\n",
      "    387        raise cls(str(exc)) from exc\n",
      "    388\n",
      "    389    def diagnostics(self) -> str:\n",
      "    390        \"\"\"Returns diagnostics for the error.\"\"\"\n",
      "    391        if isinstance(self.__cause__, id.GitHubOidcPermissionCredentialError):\n",
      "    392            return f\"\"\"\n",
      "    393                Insufficient permissions for GitHub Actions workflow.\n",
      "    394\n",
      "    395                The most common reason for this is incorrect\n",
      "    396                configuration of the top-level `permissions` setting of the\n",
      "    397                workflow YAML file. It should be configured like so:\n",
      "    398\n",
      "    399                    permissions:\n",
      "    400                      id-token: write\n",
      "    401\n",
      "    402                Relevant documentation here:\n",
      "    403\n",
      "    404                    https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#adding-permissions-settings\n",
      "    405\n",
      "    406                Another possible reason is that the workflow run has been\n",
      "    407                triggered by a PR from a forked repository. PRs from forked\n",
      "    408                repositories typically cannot be granted write access.\n",
      "    409\n",
      "    410                Relevant documentation here:\n",
      "    411\n",
      "    412                    https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token\n",
      "    413\n",
      "    414                Additional context:\n",
      "    415\n",
      "    416                {self.__cause__}\n",
      "    417                \"\"\"\n",
      "    418        else:\n",
      "    419            return f\"\"\"\n",
      "    420                An issue occurred with ambient credential detection.\n",
      "    421\n",
      "    422                Additional context:\n",
      "    423\n",
      "    424                {self}\n",
      "    425            \"\"\"\n",
      "    426\n",
      "    427\n",
      "    428def detect_credential() -> Optional[str]:\n",
      "    429    \"\"\"Calls `id.detect_credential`, but wraps exceptions with our own exception type.\"\"\"\n",
      "    430    try:\n",
      "    431        return cast(Optional[str], id.detect_credential(_DEFAULT_AUDIENCE))\n",
      "    432    except id.IdentityError as exc:\n",
      "    433        IdentityError.raise_from_id(exc)\n",
      "    \n",
      "\n",
      "DEFAULT_OAUTH_ISSUER_URL = 'https://oauth2.sigstore.dev/auth'\n",
      "\n",
      "STAGING_OAUTH_ISSUER_URL = 'https://oauth2.sigstage.dev/auth'\n",
      "\n",
      "DEFAULT_AUDIENCE = 'sigstore'\n",
      "\n",
      "class ExpiredIdentity(builtins.Exception): View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    72class ExpiredIdentity(Exception):\n",
      "    73    \"\"\"An error raised when an identity token is expired.\"\"\"\n",
      "    \n",
      "\n",
      "An error raised when an identity token is expired.\n",
      "\n",
      "##### Inherited Members\n",
      "\n",
      "builtins.Exception\n",
      "\n",
      "    Exception\n",
      "\n",
      "builtins.BaseException\n",
      "\n",
      "    with_traceback\n",
      "    add_note\n",
      "    args\n",
      "\n",
      "class IdentityToken: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "     76class IdentityToken:\n",
      "     77    \"\"\"\n",
      "     78    An OIDC \"identity\", corresponding to an underlying OIDC token with\n",
      "     79    a sensible subject, issuer, and audience for Sigstore purposes.\n",
      "     80    \"\"\"\n",
      "     81\n",
      "     82    def __init__(self, raw_token: str) -> None:\n",
      "     83        \"\"\"\n",
      "     84        Create a new `IdentityToken` from the given OIDC token.\n",
      "     85        \"\"\"\n",
      "     86\n",
      "     87        self._raw_token = raw_token\n",
      "     88\n",
      "     89        # NOTE: The lack of verification here is intentional, and is part of\n",
      "     90        # Sigstore's verification model: clients like sigstore-python are\n",
      "     91        # responsible only for forwarding the OIDC identity to Fulcio for\n",
      "     92        # certificate binding and issuance.\n",
      "     93        try:\n",
      "     94            self._unverified_claims = jwt.decode(\n",
      "     95                raw_token,\n",
      "     96                options={\n",
      "     97                    \"verify_signature\": False,\n",
      "     98                    \"verify_aud\": True,\n",
      "     99                    \"verify_iat\": True,\n",
      "    100                    \"verify_exp\": True,\n",
      "    101                    # These claims are required by OpenID Connect, so\n",
      "    102                    # we can strongly enforce their presence.\n",
      "    103                    # See: https://openid.net/specs/openid-connect-basic-1_0.html#IDToken\n",
      "    104                    \"require\": [\"aud\", \"sub\", \"iat\", \"exp\", \"iss\"],\n",
      "    105                },\n",
      "    106                audience=DEFAULT_AUDIENCE,\n",
      "    107                # NOTE: This leeway shouldn't be strictly necessary, but is\n",
      "    108                # included to preempt any (small) skew between the host\n",
      "    109                # and the originating IdP.\n",
      "    110                leeway=5,\n",
      "    111            )\n",
      "    112        except Exception as exc:\n",
      "    113            raise IdentityError(\n",
      "    114                \"Identity token is malformed or missing claims\"\n",
      "    115            ) from exc\n",
      "    116\n",
      "    117        self._iss: str = self._unverified_claims[\"iss\"]\n",
      "    118        self._nbf: int | None = self._unverified_claims.get(\"nbf\")\n",
      "    119        self._exp: int = self._unverified_claims[\"exp\"]\n",
      "    120\n",
      "    121        # Fail early if this token isn't within its validity period.\n",
      "    122        if not self.in_validity_period():\n",
      "    123            raise IdentityError(\"Identity token is not within its validity period\")\n",
      "    124\n",
      "    125        # When verifying the private key possession proof, Fulcio uses\n",
      "    126        # different claims depending on the token's issuer.\n",
      "    127        # We currently special-case a handful of these, and fall back\n",
      "    128        # on signing the \"sub\" claim otherwise.\n",
      "    129        identity_claim = _KNOWN_OIDC_ISSUERS.get(self.issuer)\n",
      "    130        if identity_claim is not None:\n",
      "    131            if identity_claim not in self._unverified_claims:\n",
      "    132                raise IdentityError(\n",
      "    133                    f\"Identity token is missing the required {identity_claim!r} claim\"\n",
      "    134                )\n",
      "    135\n",
      "    136            self._identity = str(self._unverified_claims.get(identity_claim))\n",
      "    137        else:\n",
      "    138            try:\n",
      "    139                self._identity = str(self._unverified_claims[\"sub\"])\n",
      "    140            except KeyError:\n",
      "    141                raise IdentityError(\n",
      "    142                    \"Identity token is missing the required 'sub' claim\"\n",
      "    143                )\n",
      "    144\n",
      "    145        # This identity token might have been retrieved directly from\n",
      "    146        # an identity provider, or it might be a \"federated\" identity token\n",
      "    147        # retrieved from a federated IdP (e.g., Sigstore's own Dex instance).\n",
      "    148        # In the latter case, the claims will also include a `federated_claims`\n",
      "    149        # set, which in turn should include a `connector_id` that reflects\n",
      "    150        # the \"real\" token issuer. We retrieve this, despite technically\n",
      "    151        # being an implementation detail, because it has value to client\n",
      "    152        # users: a client might want to make sure that its user is identifying\n",
      "    153        # with a *particular* IdP, which means that they need to pierce the\n",
      "    154        # federation layer to check which IdP is actually being used.\n",
      "    155        self._federated_issuer: str | None = None\n",
      "    156        federated_claims = self._unverified_claims.get(\"federated_claims\")\n",
      "    157        if federated_claims is not None:\n",
      "    158            if not isinstance(federated_claims, dict):\n",
      "    159                raise IdentityError(\n",
      "    160                    \"unexpected claim type: federated_claims is not a dict\"\n",
      "    161                )\n",
      "    162\n",
      "    163            federated_issuer = federated_claims.get(\"connector_id\")\n",
      "    164            if federated_issuer is not None:\n",
      "    165                if not isinstance(federated_issuer, str):\n",
      "    166                    raise IdentityError(\n",
      "    167                        \"unexpected claim type: federated_claims.connector_id is not a string\"\n",
      "    168                    )\n",
      "    169\n",
      "    170                self._federated_issuer = federated_issuer\n",
      "    171\n",
      "    172    def in_validity_period(self) -> bool:\n",
      "    173        \"\"\"\n",
      "    174        Returns whether or not this `Identity` is currently within its self-stated validity period.\n",
      "    175\n",
      "    176        NOTE: As noted in `Identity.__init__`, this is not a verifying wrapper;\n",
      "    177        the check here only asserts whether the *unverified* identity's claims\n",
      "    178        are within their validity period.\n",
      "    179        \"\"\"\n",
      "    180\n",
      "    181        now = datetime.now(timezone.utc).timestamp()\n",
      "    182\n",
      "    183        if self._nbf is not None:\n",
      "    184            return self._nbf <= now < self._exp\n",
      "    185        else:\n",
      "    186            return now < self._exp\n",
      "    187\n",
      "    188    @property\n",
      "    189    def identity(self) -> str:\n",
      "    190        \"\"\"\n",
      "    191        Returns this `IdentityToken`'s underlying \"subject\".\n",
      "    192\n",
      "    193        Note that this is **not** always the `sub` claim in the corresponding\n",
      "    194        identity token: depending onm the token's issuer, it may be a *different*\n",
      "    195        claim, such as `email`. This corresponds to the Sigstore ecosystem's\n",
      "    196        behavior, e.g. in each issued certificate's SAN.\n",
      "    197        \"\"\"\n",
      "    198        return self._identity\n",
      "    199\n",
      "    200    @property\n",
      "    201    def issuer(self) -> str:\n",
      "    202        \"\"\"\n",
      "    203        Returns a URL identifying this `IdentityToken`'s issuer.\n",
      "    204        \"\"\"\n",
      "    205        return self._iss\n",
      "    206\n",
      "    207    @property\n",
      "    208    def expected_certificate_subject(self) -> str:\n",
      "    209        \"\"\"\n",
      "    210        Returns a URL identifying the **expected** subject for any Sigstore\n",
      "    211        certificate issued against this identity token.\n",
      "    212\n",
      "    213        The behavior of this field is slightly subtle: for non-federated\n",
      "    214        identity providers (like a token issued directly by Google's IdP) it\n",
      "    215        should be exactly equivalent to `IdentityToken.issuer`. For federated\n",
      "    216        issuers (like Sigstore's own federated IdP) it should be equivalent to\n",
      "    217        the underlying federated issuer's URL, which is kept in an\n",
      "    218        implementation-defined claim.\n",
      "    219\n",
      "    220        This attribute exists so that clients who wish to inspect the expected\n",
      "    221        subject of their certificates can do so without relying on\n",
      "    222        implementation-specific behavior.\n",
      "    223        \"\"\"\n",
      "    224        if self._federated_issuer is not None:\n",
      "    225            return self._federated_issuer\n",
      "    226\n",
      "    227        return self.issuer\n",
      "    228\n",
      "    229    def __str__(self) -> str:\n",
      "    230        \"\"\"\n",
      "    231        Returns the underlying OIDC token for this identity.\n",
      "    232\n",
      "    233        That this token is secret in nature and **MUST NOT** be disclosed.\n",
      "    234        \"\"\"\n",
      "    235        return self._raw_token\n",
      "    \n",
      "\n",
      "An OIDC \"identity\", corresponding to an underlying OIDC token with a sensible\n",
      "subject, issuer, and audience for Sigstore purposes.\n",
      "\n",
      "IdentityToken(raw_token: str) View Source\n",
      "\n",
      "    \n",
      "    \n",
      "     82    def __init__(self, raw_token: str) -> None:\n",
      "     83        \"\"\"\n",
      "     84        Create a new `IdentityToken` from the given OIDC token.\n",
      "     85        \"\"\"\n",
      "     86\n",
      "     87        self._raw_token = raw_token\n",
      "     88\n",
      "     89        # NOTE: The lack of verification here is intentional, and is part of\n",
      "     90        # Sigstore's verification model: clients like sigstore-python are\n",
      "     91        # responsible only for forwarding the OIDC identity to Fulcio for\n",
      "     92        # certificate binding and issuance.\n",
      "     93        try:\n",
      "     94            self._unverified_claims = jwt.decode(\n",
      "     95                raw_token,\n",
      "     96                options={\n",
      "     97                    \"verify_signature\": False,\n",
      "     98                    \"verify_aud\": True,\n",
      "     99                    \"verify_iat\": True,\n",
      "    100                    \"verify_exp\": True,\n",
      "    101                    # These claims are required by OpenID Connect, so\n",
      "    102                    # we can strongly enforce their presence.\n",
      "    103                    # See: https://openid.net/specs/openid-connect-basic-1_0.html#IDToken\n",
      "    104                    \"require\": [\"aud\", \"sub\", \"iat\", \"exp\", \"iss\"],\n",
      "    105                },\n",
      "    106                audience=DEFAULT_AUDIENCE,\n",
      "    107                # NOTE: This leeway shouldn't be strictly necessary, but is\n",
      "    108                # included to preempt any (small) skew between the host\n",
      "    109                # and the originating IdP.\n",
      "    110                leeway=5,\n",
      "    111            )\n",
      "    112        except Exception as exc:\n",
      "    113            raise IdentityError(\n",
      "    114                \"Identity token is malformed or missing claims\"\n",
      "    115            ) from exc\n",
      "    116\n",
      "    117        self._iss: str = self._unverified_claims[\"iss\"]\n",
      "    118        self._nbf: int | None = self._unverified_claims.get(\"nbf\")\n",
      "    119        self._exp: int = self._unverified_claims[\"exp\"]\n",
      "    120\n",
      "    121        # Fail early if this token isn't within its validity period.\n",
      "    122        if not self.in_validity_period():\n",
      "    123            raise IdentityError(\"Identity token is not within its validity period\")\n",
      "    124\n",
      "    125        # When verifying the private key possession proof, Fulcio uses\n",
      "    126        # different claims depending on the token's issuer.\n",
      "    127        # We currently special-case a handful of these, and fall back\n",
      "    128        # on signing the \"sub\" claim otherwise.\n",
      "    129        identity_claim = _KNOWN_OIDC_ISSUERS.get(self.issuer)\n",
      "    130        if identity_claim is not None:\n",
      "    131            if identity_claim not in self._unverified_claims:\n",
      "    132                raise IdentityError(\n",
      "    133                    f\"Identity token is missing the required {identity_claim!r} claim\"\n",
      "    134                )\n",
      "    135\n",
      "    136            self._identity = str(self._unverified_claims.get(identity_claim))\n",
      "    137        else:\n",
      "    138            try:\n",
      "    139                self._identity = str(self._unverified_claims[\"sub\"])\n",
      "    140            except KeyError:\n",
      "    141                raise IdentityError(\n",
      "    142                    \"Identity token is missing the required 'sub' claim\"\n",
      "    143                )\n",
      "    144\n",
      "    145        # This identity token might have been retrieved directly from\n",
      "    146        # an identity provider, or it might be a \"federated\" identity token\n",
      "    147        # retrieved from a federated IdP (e.g., Sigstore's own Dex instance).\n",
      "    148        # In the latter case, the claims will also include a `federated_claims`\n",
      "    149        # set, which in turn should include a `connector_id` that reflects\n",
      "    150        # the \"real\" token issuer. We retrieve this, despite technically\n",
      "    151        # being an implementation detail, because it has value to client\n",
      "    152        # users: a client might want to make sure that its user is identifying\n",
      "    153        # with a *particular* IdP, which means that they need to pierce the\n",
      "    154        # federation layer to check which IdP is actually being used.\n",
      "    155        self._federated_issuer: str | None = None\n",
      "    156        federated_claims = self._unverified_claims.get(\"federated_claims\")\n",
      "    157        if federated_claims is not None:\n",
      "    158            if not isinstance(federated_claims, dict):\n",
      "    159                raise IdentityError(\n",
      "    160                    \"unexpected claim type: federated_claims is not a dict\"\n",
      "    161                )\n",
      "    162\n",
      "    163            federated_issuer = federated_claims.get(\"connector_id\")\n",
      "    164            if federated_issuer is not None:\n",
      "    165                if not isinstance(federated_issuer, str):\n",
      "    166                    raise IdentityError(\n",
      "    167                        \"unexpected claim type: federated_claims.connector_id is not a string\"\n",
      "    168                    )\n",
      "    169\n",
      "    170                self._federated_issuer = federated_issuer\n",
      "    \n",
      "\n",
      "Create a new `IdentityToken` from the given OIDC token.\n",
      "\n",
      "def in_validity_period(self) -> bool: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    172    def in_validity_period(self) -> bool:\n",
      "    173        \"\"\"\n",
      "    174        Returns whether or not this `Identity` is currently within its self-stated validity period.\n",
      "    175\n",
      "    176        NOTE: As noted in `Identity.__init__`, this is not a verifying wrapper;\n",
      "    177        the check here only asserts whether the *unverified* identity's claims\n",
      "    178        are within their validity period.\n",
      "    179        \"\"\"\n",
      "    180\n",
      "    181        now = datetime.now(timezone.utc).timestamp()\n",
      "    182\n",
      "    183        if self._nbf is not None:\n",
      "    184            return self._nbf <= now < self._exp\n",
      "    185        else:\n",
      "    186            return now < self._exp\n",
      "    \n",
      "\n",
      "Returns whether or not this `Identity` is currently within its self-stated\n",
      "validity period.\n",
      "\n",
      "NOTE: As noted in `Identity.__init__`, this is not a verifying wrapper; the\n",
      "check here only asserts whether the _unverified_ identity's claims are within\n",
      "their validity period.\n",
      "\n",
      "identity: str\n",
      "\n",
      "Returns this `IdentityToken`'s underlying \"subject\".\n",
      "\n",
      "Note that this is **not** always the `sub` claim in the corresponding identity\n",
      "token: depending onm the token's issuer, it may be a _different_ claim, such\n",
      "as `email`. This corresponds to the Sigstore ecosystem's behavior, e.g. in\n",
      "each issued certificate's SAN.\n",
      "\n",
      "issuer: str\n",
      "\n",
      "Returns a URL identifying this `IdentityToken`'s issuer.\n",
      "\n",
      "expected_certificate_subject: str\n",
      "\n",
      "Returns a URL identifying the **expected** subject for any Sigstore\n",
      "certificate issued against this identity token.\n",
      "\n",
      "The behavior of this field is slightly subtle: for non-federated identity\n",
      "providers (like a token issued directly by Google's IdP) it should be exactly\n",
      "equivalent to `IdentityToken.issuer`. For federated issuers (like Sigstore's\n",
      "own federated IdP) it should be equivalent to the underlying federated\n",
      "issuer's URL, which is kept in an implementation-defined claim.\n",
      "\n",
      "This attribute exists so that clients who wish to inspect the expected subject\n",
      "of their certificates can do so without relying on implementation-specific\n",
      "behavior.\n",
      "\n",
      "class IssuerError(builtins.Exception): View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    238class IssuerError(Exception):\n",
      "    239    \"\"\"\n",
      "    240    Raised on any communication or format error with an OIDC issuer.\n",
      "    241    \"\"\"\n",
      "    242\n",
      "    243    pass\n",
      "    \n",
      "\n",
      "Raised on any communication or format error with an OIDC issuer.\n",
      "\n",
      "##### Inherited Members\n",
      "\n",
      "builtins.Exception\n",
      "\n",
      "    Exception\n",
      "\n",
      "builtins.BaseException\n",
      "\n",
      "    with_traceback\n",
      "    add_note\n",
      "    args\n",
      "\n",
      "class Issuer: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    246class Issuer:\n",
      "    247    \"\"\"\n",
      "    248    Represents an OIDC issuer (IdP).\n",
      "    249    \"\"\"\n",
      "    250\n",
      "    251    def __init__(self, base_url: str) -> None:\n",
      "    252        \"\"\"\n",
      "    253        Create a new `Issuer` from the given base URL.\n",
      "    254\n",
      "    255        This URL is used to locate an OpenID Connect configuration file,\n",
      "    256        which is then used to bootstrap the issuer's state (such\n",
      "    257        as authorization and token endpoints).\n",
      "    258        \"\"\"\n",
      "    259        oidc_config_url = urllib.parse.urljoin(\n",
      "    260            f\"{base_url}/\", \".well-known/openid-configuration\"\n",
      "    261        )\n",
      "    262\n",
      "    263        try:\n",
      "    264            resp: requests.Response = requests.get(oidc_config_url, timeout=30)\n",
      "    265        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "    266            raise NetworkError from exc\n",
      "    267\n",
      "    268        try:\n",
      "    269            resp.raise_for_status()\n",
      "    270        except requests.HTTPError as http_error:\n",
      "    271            raise IssuerError from http_error\n",
      "    272\n",
      "    273        try:\n",
      "    274            # We don't generally expect this to fail (since the provider should\n",
      "    275            # return a non-success HTTP code which we catch above), but we\n",
      "    276            # check just in case we have a misbehaving OIDC issuer.\n",
      "    277            self.oidc_config = _OpenIDConfiguration.model_validate(resp.json())\n",
      "    278        except ValueError as exc:\n",
      "    279            raise IssuerError(f\"OIDC issuer returned invalid configuration: {exc}\")\n",
      "    280\n",
      "    281    @classmethod\n",
      "    282    def production(cls) -> Issuer:\n",
      "    283        \"\"\"\n",
      "    284        Returns an `Issuer` configured against Sigstore's production-level services.\n",
      "    285        \"\"\"\n",
      "    286        return cls(DEFAULT_OAUTH_ISSUER_URL)\n",
      "    287\n",
      "    288    @classmethod\n",
      "    289    def staging(cls) -> Issuer:\n",
      "    290        \"\"\"\n",
      "    291        Returns an `Issuer` configured against Sigstore's staging-level services.\n",
      "    292        \"\"\"\n",
      "    293        return cls(STAGING_OAUTH_ISSUER_URL)\n",
      "    294\n",
      "    295    def identity_token(  # nosec: B107\n",
      "    296        self,\n",
      "    297        client_id: str = \"sigstore\",\n",
      "    298        client_secret: str = \"\",\n",
      "    299        force_oob: bool = False,\n",
      "    300    ) -> IdentityToken:\n",
      "    301        \"\"\"\n",
      "    302        Retrieves and returns an `IdentityToken` from the current `Issuer`, via OAuth.\n",
      "    303\n",
      "    304        This function blocks on user interaction.\n",
      "    305\n",
      "    306        The `force_oob` flag controls the kind of flow performed. When `False` (the default),\n",
      "    307        this function attempts to open the user's web browser before falling back to\n",
      "    308        an out-of-band flow. When `True`, the out-of-band flow is always used.\n",
      "    309        \"\"\"\n",
      "    310\n",
      "    311        # This function and the components that it relies on are based off of:\n",
      "    312        # https://github.com/psteniusubi/python-sample\n",
      "    313\n",
      "    314        from sigstore._internal.oidc.oauth import _OAuthFlow\n",
      "    315\n",
      "    316        code: str\n",
      "    317        with _OAuthFlow(client_id, client_secret, self) as server:\n",
      "    318            # Launch web browser\n",
      "    319            if not force_oob and webbrowser.open(server.base_uri):\n",
      "    320                print(\"Waiting for browser interaction...\", file=sys.stderr)\n",
      "    321            else:\n",
      "    322                server.enable_oob()\n",
      "    323                print(\n",
      "    324                    f\"Go to the following link in a browser:\\n\\n\\t{server.auth_endpoint}\",\n",
      "    325                    file=sys.stderr,\n",
      "    326                )\n",
      "    327\n",
      "    328            if not server.is_oob():\n",
      "    329                # Wait until the redirect server populates the response\n",
      "    330                while server.auth_response is None:\n",
      "    331                    time.sleep(0.1)\n",
      "    332\n",
      "    333                auth_error = server.auth_response.get(\"error\")\n",
      "    334                if auth_error is not None:\n",
      "    335                    raise IdentityError(\n",
      "    336                        f\"Error response from auth endpoint: {auth_error[0]}\"\n",
      "    337                    )\n",
      "    338                code = server.auth_response[\"code\"][0]\n",
      "    339            else:\n",
      "    340                # In the out-of-band case, we wait until the user provides the code\n",
      "    341                code = input(\"Enter verification code: \")\n",
      "    342\n",
      "    343        # Provide code to token endpoint\n",
      "    344        data = {\n",
      "    345            \"grant_type\": \"authorization_code\",\n",
      "    346            \"redirect_uri\": server.redirect_uri,\n",
      "    347            \"code\": code,\n",
      "    348            \"code_verifier\": server.oauth_session.code_verifier,\n",
      "    349        }\n",
      "    350        auth = (\n",
      "    351            client_id,\n",
      "    352            client_secret,\n",
      "    353        )\n",
      "    354        logging.debug(f\"PAYLOAD: data={data}\")\n",
      "    355        try:\n",
      "    356            resp: requests.Response = requests.post(\n",
      "    357                self.oidc_config.token_endpoint,\n",
      "    358                data=data,\n",
      "    359                auth=auth,\n",
      "    360                timeout=30,\n",
      "    361            )\n",
      "    362        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "    363            raise NetworkError from exc\n",
      "    364\n",
      "    365        try:\n",
      "    366            resp.raise_for_status()\n",
      "    367        except requests.HTTPError as http_error:\n",
      "    368            raise IdentityError(\n",
      "    369                f\"Token request failed with {resp.status_code}\"\n",
      "    370            ) from http_error\n",
      "    371\n",
      "    372        token_json = resp.json()\n",
      "    373        token_error = token_json.get(\"error\")\n",
      "    374        if token_error is not None:\n",
      "    375            raise IdentityError(f\"Error response from token endpoint: {token_error}\")\n",
      "    376\n",
      "    377        return IdentityToken(token_json[\"access_token\"])\n",
      "    \n",
      "\n",
      "Represents an OIDC issuer (IdP).\n",
      "\n",
      "Issuer(base_url: str) View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    251    def __init__(self, base_url: str) -> None:\n",
      "    252        \"\"\"\n",
      "    253        Create a new `Issuer` from the given base URL.\n",
      "    254\n",
      "    255        This URL is used to locate an OpenID Connect configuration file,\n",
      "    256        which is then used to bootstrap the issuer's state (such\n",
      "    257        as authorization and token endpoints).\n",
      "    258        \"\"\"\n",
      "    259        oidc_config_url = urllib.parse.urljoin(\n",
      "    260            f\"{base_url}/\", \".well-known/openid-configuration\"\n",
      "    261        )\n",
      "    262\n",
      "    263        try:\n",
      "    264            resp: requests.Response = requests.get(oidc_config_url, timeout=30)\n",
      "    265        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "    266            raise NetworkError from exc\n",
      "    267\n",
      "    268        try:\n",
      "    269            resp.raise_for_status()\n",
      "    270        except requests.HTTPError as http_error:\n",
      "    271            raise IssuerError from http_error\n",
      "    272\n",
      "    273        try:\n",
      "    274            # We don't generally expect this to fail (since the provider should\n",
      "    275            # return a non-success HTTP code which we catch above), but we\n",
      "    276            # check just in case we have a misbehaving OIDC issuer.\n",
      "    277            self.oidc_config = _OpenIDConfiguration.model_validate(resp.json())\n",
      "    278        except ValueError as exc:\n",
      "    279            raise IssuerError(f\"OIDC issuer returned invalid configuration: {exc}\")\n",
      "    \n",
      "\n",
      "Create a new `Issuer` from the given base URL.\n",
      "\n",
      "This URL is used to locate an OpenID Connect configuration file, which is then\n",
      "used to bootstrap the issuer's state (such as authorization and token\n",
      "endpoints).\n",
      "\n",
      "@classmethod\n",
      "\n",
      "def production(cls) -> Issuer: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    281    @classmethod\n",
      "    282    def production(cls) -> Issuer:\n",
      "    283        \"\"\"\n",
      "    284        Returns an `Issuer` configured against Sigstore's production-level services.\n",
      "    285        \"\"\"\n",
      "    286        return cls(DEFAULT_OAUTH_ISSUER_URL)\n",
      "    \n",
      "\n",
      "Returns an `Issuer` configured against Sigstore's production-level services.\n",
      "\n",
      "@classmethod\n",
      "\n",
      "def staging(cls) -> Issuer: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    288    @classmethod\n",
      "    289    def staging(cls) -> Issuer:\n",
      "    290        \"\"\"\n",
      "    291        Returns an `Issuer` configured against Sigstore's staging-level services.\n",
      "    292        \"\"\"\n",
      "    293        return cls(STAGING_OAUTH_ISSUER_URL)\n",
      "    \n",
      "\n",
      "Returns an `Issuer` configured against Sigstore's staging-level services.\n",
      "\n",
      "def identity_token( self, client_id: str = 'sigstore', client_secret: str =\n",
      "'', force_oob: bool = False) -> IdentityToken: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    295    def identity_token(  # nosec: B107\n",
      "    296        self,\n",
      "    297        client_id: str = \"sigstore\",\n",
      "    298        client_secret: str = \"\",\n",
      "    299        force_oob: bool = False,\n",
      "    300    ) -> IdentityToken:\n",
      "    301        \"\"\"\n",
      "    302        Retrieves and returns an `IdentityToken` from the current `Issuer`, via OAuth.\n",
      "    303\n",
      "    304        This function blocks on user interaction.\n",
      "    305\n",
      "    306        The `force_oob` flag controls the kind of flow performed. When `False` (the default),\n",
      "    307        this function attempts to open the user's web browser before falling back to\n",
      "    308        an out-of-band flow. When `True`, the out-of-band flow is always used.\n",
      "    309        \"\"\"\n",
      "    310\n",
      "    311        # This function and the components that it relies on are based off of:\n",
      "    312        # https://github.com/psteniusubi/python-sample\n",
      "    313\n",
      "    314        from sigstore._internal.oidc.oauth import _OAuthFlow\n",
      "    315\n",
      "    316        code: str\n",
      "    317        with _OAuthFlow(client_id, client_secret, self) as server:\n",
      "    318            # Launch web browser\n",
      "    319            if not force_oob and webbrowser.open(server.base_uri):\n",
      "    320                print(\"Waiting for browser interaction...\", file=sys.stderr)\n",
      "    321            else:\n",
      "    322                server.enable_oob()\n",
      "    323                print(\n",
      "    324                    f\"Go to the following link in a browser:\\n\\n\\t{server.auth_endpoint}\",\n",
      "    325                    file=sys.stderr,\n",
      "    326                )\n",
      "    327\n",
      "    328            if not server.is_oob():\n",
      "    329                # Wait until the redirect server populates the response\n",
      "    330                while server.auth_response is None:\n",
      "    331                    time.sleep(0.1)\n",
      "    332\n",
      "    333                auth_error = server.auth_response.get(\"error\")\n",
      "    334                if auth_error is not None:\n",
      "    335                    raise IdentityError(\n",
      "    336                        f\"Error response from auth endpoint: {auth_error[0]}\"\n",
      "    337                    )\n",
      "    338                code = server.auth_response[\"code\"][0]\n",
      "    339            else:\n",
      "    340                # In the out-of-band case, we wait until the user provides the code\n",
      "    341                code = input(\"Enter verification code: \")\n",
      "    342\n",
      "    343        # Provide code to token endpoint\n",
      "    344        data = {\n",
      "    345            \"grant_type\": \"authorization_code\",\n",
      "    346            \"redirect_uri\": server.redirect_uri,\n",
      "    347            \"code\": code,\n",
      "    348            \"code_verifier\": server.oauth_session.code_verifier,\n",
      "    349        }\n",
      "    350        auth = (\n",
      "    351            client_id,\n",
      "    352            client_secret,\n",
      "    353        )\n",
      "    354        logging.debug(f\"PAYLOAD: data={data}\")\n",
      "    355        try:\n",
      "    356            resp: requests.Response = requests.post(\n",
      "    357                self.oidc_config.token_endpoint,\n",
      "    358                data=data,\n",
      "    359                auth=auth,\n",
      "    360                timeout=30,\n",
      "    361            )\n",
      "    362        except (requests.ConnectionError, requests.Timeout) as exc:\n",
      "    363            raise NetworkError from exc\n",
      "    364\n",
      "    365        try:\n",
      "    366            resp.raise_for_status()\n",
      "    367        except requests.HTTPError as http_error:\n",
      "    368            raise IdentityError(\n",
      "    369                f\"Token request failed with {resp.status_code}\"\n",
      "    370            ) from http_error\n",
      "    371\n",
      "    372        token_json = resp.json()\n",
      "    373        token_error = token_json.get(\"error\")\n",
      "    374        if token_error is not None:\n",
      "    375            raise IdentityError(f\"Error response from token endpoint: {token_error}\")\n",
      "    376\n",
      "    377        return IdentityToken(token_json[\"access_token\"])\n",
      "    \n",
      "\n",
      "Retrieves and returns an `IdentityToken` from the current `Issuer`, via OAuth.\n",
      "\n",
      "This function blocks on user interaction.\n",
      "\n",
      "The `force_oob` flag controls the kind of flow performed. When `False` (the\n",
      "default), this function attempts to open the user's web browser before falling\n",
      "back to an out-of-band flow. When `True`, the out-of-band flow is always used.\n",
      "\n",
      "class IdentityError([sigstore.errors.Error](errors.html#Error)): View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    380class IdentityError(Error):\n",
      "    381    \"\"\"\n",
      "    382    Wraps `id`'s IdentityError.\n",
      "    383    \"\"\"\n",
      "    384\n",
      "    385    @classmethod\n",
      "    386    def raise_from_id(cls, exc: id.IdentityError) -> NoReturn:\n",
      "    387        \"\"\"Raises a wrapped IdentityError from the provided `id.IdentityError`.\"\"\"\n",
      "    388        raise cls(str(exc)) from exc\n",
      "    389\n",
      "    390    def diagnostics(self) -> str:\n",
      "    391        \"\"\"Returns diagnostics for the error.\"\"\"\n",
      "    392        if isinstance(self.__cause__, id.GitHubOidcPermissionCredentialError):\n",
      "    393            return f\"\"\"\n",
      "    394                Insufficient permissions for GitHub Actions workflow.\n",
      "    395\n",
      "    396                The most common reason for this is incorrect\n",
      "    397                configuration of the top-level `permissions` setting of the\n",
      "    398                workflow YAML file. It should be configured like so:\n",
      "    399\n",
      "    400                    permissions:\n",
      "    401                      id-token: write\n",
      "    402\n",
      "    403                Relevant documentation here:\n",
      "    404\n",
      "    405                    https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#adding-permissions-settings\n",
      "    406\n",
      "    407                Another possible reason is that the workflow run has been\n",
      "    408                triggered by a PR from a forked repository. PRs from forked\n",
      "    409                repositories typically cannot be granted write access.\n",
      "    410\n",
      "    411                Relevant documentation here:\n",
      "    412\n",
      "    413                    https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token\n",
      "    414\n",
      "    415                Additional context:\n",
      "    416\n",
      "    417                {self.__cause__}\n",
      "    418                \"\"\"\n",
      "    419        else:\n",
      "    420            return f\"\"\"\n",
      "    421                An issue occurred with ambient credential detection.\n",
      "    422\n",
      "    423                Additional context:\n",
      "    424\n",
      "    425                {self}\n",
      "    426            \"\"\"\n",
      "    \n",
      "\n",
      "Wraps `id`'s IdentityError.\n",
      "\n",
      "@classmethod\n",
      "\n",
      "def raise_from_id(cls, exc: id.IdentityError) -> NoReturn: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    385    @classmethod\n",
      "    386    def raise_from_id(cls, exc: id.IdentityError) -> NoReturn:\n",
      "    387        \"\"\"Raises a wrapped IdentityError from the provided `id.IdentityError`.\"\"\"\n",
      "    388        raise cls(str(exc)) from exc\n",
      "    \n",
      "\n",
      "Raises a wrapped IdentityError from the provided `id.IdentityError`.\n",
      "\n",
      "def diagnostics(self) -> str: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    390    def diagnostics(self) -> str:\n",
      "    391        \"\"\"Returns diagnostics for the error.\"\"\"\n",
      "    392        if isinstance(self.__cause__, id.GitHubOidcPermissionCredentialError):\n",
      "    393            return f\"\"\"\n",
      "    394                Insufficient permissions for GitHub Actions workflow.\n",
      "    395\n",
      "    396                The most common reason for this is incorrect\n",
      "    397                configuration of the top-level `permissions` setting of the\n",
      "    398                workflow YAML file. It should be configured like so:\n",
      "    399\n",
      "    400                    permissions:\n",
      "    401                      id-token: write\n",
      "    402\n",
      "    403                Relevant documentation here:\n",
      "    404\n",
      "    405                    https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#adding-permissions-settings\n",
      "    406\n",
      "    407                Another possible reason is that the workflow run has been\n",
      "    408                triggered by a PR from a forked repository. PRs from forked\n",
      "    409                repositories typically cannot be granted write access.\n",
      "    410\n",
      "    411                Relevant documentation here:\n",
      "    412\n",
      "    413                    https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token\n",
      "    414\n",
      "    415                Additional context:\n",
      "    416\n",
      "    417                {self.__cause__}\n",
      "    418                \"\"\"\n",
      "    419        else:\n",
      "    420            return f\"\"\"\n",
      "    421                An issue occurred with ambient credential detection.\n",
      "    422\n",
      "    423                Additional context:\n",
      "    424\n",
      "    425                {self}\n",
      "    426            \"\"\"\n",
      "    \n",
      "\n",
      "Returns diagnostics for the error.\n",
      "\n",
      "##### Inherited Members\n",
      "\n",
      "builtins.Exception\n",
      "\n",
      "    Exception\n",
      "\n",
      "[sigstore.errors.Error](errors.html#Error)\n",
      "\n",
      "    [print_and_exit](errors.html#Error.print_and_exit)\n",
      "\n",
      "builtins.BaseException\n",
      "\n",
      "    with_traceback\n",
      "    add_note\n",
      "    args\n",
      "\n",
      "def detect_credential() -> Optional[str]: View Source\n",
      "\n",
      "    \n",
      "    \n",
      "    429def detect_credential() -> Optional[str]:\n",
      "    430    \"\"\"Calls `id.detect_credential`, but wraps exceptions with our own exception type.\"\"\"\n",
      "    431    try:\n",
      "    432        return cast(Optional[str], id.detect_credential(_DEFAULT_AUDIENCE))\n",
      "    433    except id.IdentityError as exc:\n",
      "    434        IdentityError.raise_from_id(exc)\n",
      "    \n",
      "\n",
      "Calls `id.detect_credential`, but wraps exceptions with our own exception\n",
      "type.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "expected = data['oidc']['markdown']\n",
    "print(expected)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
